{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "from Bio import SeqIO\n",
    "from collections import Counter, OrderedDict\n",
    "from itertools import permutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parent directories\n",
    "\n",
    "COORDS_DIR = \"./coords\"\t\t\t# directory containing coordinate files\n",
    "EXTRACTED_DIR = \"./extracted\"\t# directory containing extracted .vcf files (Section 1)\n",
    "\n",
    "\n",
    "WINDOWS_DIR = \"./windows\"\t\t# directory to write out windows (Section 2)\n",
    "MATCHES_DIR = \"./matches\"\t\t# directory to write out matches (Section 3)\n",
    "COUNTS_DIR = \"./counts\"\t\t\t# directory to write out counts (Section 5)\n",
    "METRICS_DIR = \"./metrics\"\t\t# directory to write out metrics (Sections 7, 8)\n",
    "\n",
    "check_dir = WINDOWS_DIR, MATCHES_DIR, COUNTS_DIR, METRICS_DIR\n",
    "for directory in check_dir:\n",
    "\tif not os.path.exists(directory):\n",
    "\t\tos.mkdir(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subdirectories\n",
    "\n",
    "POS_COUNTS_SUBDIR = f\"{COUNTS_DIR}/positional\"\t\t\t# positional counts subdirectorty (Section 5)\n",
    "POSITIONAL_METRICS_SUBDIR = f\"{METRICS_DIR}/positional\"\t# positional metrics subdirectorty (Section 7)\n",
    "GENE_LEVEL_METRICS_SUBDIR = f\"{METRICS_DIR}/gene_level\"\t# gene level metrics subdirectorty (Section 8)\n",
    "\n",
    "check_dir = POS_COUNTS_SUBDIR, POSITIONAL_METRICS_SUBDIR, GENE_LEVEL_METRICS_SUBDIR\n",
    "for directory in check_dir:\n",
    "\tif not os.path.exists(directory):\n",
    "\t\tos.mkdir(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# paths to coordinate files\n",
    "TSS_COORDS_PATH = f\"{COORDS_DIR}/Fantom5_CAGE_hg38_gencode.v40_TSS_EIB_Updated_200.txt\"\n",
    "INTERGEN_GCMATCH_COORDS_PATH = f\"{COORDS_DIR}/hg38_intergenic_more_1.txt\"\n",
    "INTERGEN_NO_GCMATCH_COORDS_PATH = f\"{COORDS_DIR}/hg38_intergenic_notGCmatched.txt\"\n",
    "\n",
    "# path to genome FASTA file\n",
    "HG38_PATH = \"./Homo_sapiens.GRCh38.dna.toplevel.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "BASES_SET = {'A','C','G','T'}   # set of valid characters representing the four bases\n",
    "CPG_OFFSET = 1                  # window offset to account for CpGs at the edges of windows\n",
    "CPG_STR = 'CG'                  # string representation of CpG\n",
    "INTERGEN_GCMATCH_WINLEN = 100\t# length of intergenic GC matched window\n",
    "RANDOM_STATE  = 1337            # random state/seed for random number generation\n",
    "SLIDING_WIN_LEN = 100\t\t\t# length of sliding window used for generating metrics with a sliding window\n",
    "WINDOW_LENGTH = 5_000           # length of window\n",
    "\n",
    "# dictionary to map chromosome identifier format; \"chrom#\" -> # (i.e. \"chrom9\" -> 9)\n",
    "CHROM_DICT = { f\"chr{i}\":i\n",
    "\t\t\t\tfor i\n",
    "\t\t\t\tin range(1,24) }\n",
    "CHROM_DICT.update({\n",
    "\t'chrX':'X',\n",
    "\t'chrY':'Y',\n",
    "\t'chrM':'M',\n",
    "})\n",
    "\n",
    "# set of possible mutations (i.e. A>G, C>T, etc.)\n",
    "DNM_TYPES = { '>'.join(ref_alt)\n",
    "\t\t\t\tfor ref_alt\n",
    "\t\t\t\tin (list(permutations(BASES_SET, 2))) }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Global writing mode option. This sets the behavior when writing results out to disk.\n",
    "# There are two options:\n",
    "# 'w' = create new file if doesn't exist, or overwrite file exist\n",
    "# 'x' = create new file if doesn't exist, but does NOT overwrite file if exists\n",
    "\n",
    "WRITING_MODE = 'x'"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___________"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **1.** <u>Iceland dataset</u>"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **1.1** Download compressed files\n",
    "\n",
    "This dataset is at https://www.ebi.ac.uk/ena/browser/view/PRJEB21300. The *Download All* link downloads a shell script that downloads the *.vcf.gz* (compressed DNM information) and *.vcf.gz.tbi* (supporting information on the whole dataset; not useful for our purposes) files for each proband. To only download the *.vcf.gz* files, the cell below strips out the command to download the *.vcf.gz.tbi* files.\n",
    "\n",
    "*Note: the filename (i.e. \"ena-file-download-20230531-0107.sh\") is timestamped, thus its name is dependent on the time it was downloaded. Change this to the name of the file downloaded by you.*\n",
    "\n",
    "*Note: this shell script is easy to run on Linux but must be given executable permissions befor it can be run.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shellscript_path = \"./ena-file-download-20230531-0107.sh\"    # CHANGE THIS LINE TO POINT TO THE FILE YOU DOWNLOADED\n",
    "\n",
    "with open(f\"{shellscript_path}\", 'r') as f:\n",
    "\tlines_read = [ line\n",
    "\t\t\t\t\tfor line\n",
    "\t\t\t\t\tin f.readlines()\n",
    "\t\t\t\t\tif not line.endswith(\".tbi\\n\") ]\n",
    "\n",
    "with open(f\"{shellscript_path.strip('.sh')}_stripped.sh\", WRITING_MODE) as f:\n",
    "\tf.writelines(lines_read)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After running the script and downloading the compressed files, the files were manually extracted to `EXTRACTED_DIR`. The following section cleans up the files and converts them to comma separated files (*.csv*)."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **1.2** Read *.vcf* files"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-section reads the files into a single list, creates a Pandas DataFrame from the list, and writes the dataset to a file called \"iceland_dataset.csv\" in a comma-separated format.\n",
    "\n",
    "Please note, this does not preserve proband information. The only information that is preserved is individual DNM information. That is, only the following DNM data is kept from the original *.vcf* files:\n",
    "\n",
    "1. chromosome (\"CHROM\")\n",
    "2. position (\"POS\")\n",
    "3. reference base (\"REF\")\n",
    "4. altered base (\"ALT\")\n",
    "\n",
    "The following is ignored from each *.vcf* file:\n",
    "- comment section, i.e. lines starting with '##'\n",
    "- column headers, i.e. line starting with '#'\n",
    "- the following columns:\n",
    "\t- \"ID\"\n",
    "\t- \"QUALITY\"\n",
    "\t- \"FILTER\"\n",
    "\t- \"INFO\"\n",
    "\t- \"FORMAT\"\n",
    "\t- \"Proband-#\" (where # is the proband number)\n",
    "\n",
    "Also note: DNMs with reference or altered bases greater than 1 base in length are discarded. Only single base mutations are used for the rest of this analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse folder containing extracted .vcf files\n",
    "\n",
    "dir_list = [ f\n",
    "\t\t\t  for f\n",
    "\t\t\t  in sorted(os.listdir(EXTRACTED_DIR))\n",
    "\t\t\t  if f.endswith(\".vcf\") ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100% complete\r"
     ]
    }
   ],
   "source": [
    "# extract DNM information from each file: chromosome, position, reference base at position, base found at position in proband\n",
    "# for example:\n",
    "# > print(aggregation)\n",
    "#   [['CHROM', 'POS', 'REF', 'ALT'],\n",
    "#   [1, '2360454', 'C', 'A'],\n",
    "#               .\n",
    "#               .\n",
    "#               .\n",
    "#   [22, '26576171', 'C', 'T']]\n",
    "\n",
    "aggregation = list()\n",
    "aggregation.append([    # these will be column headers for the dataset\n",
    "\t'CHROM',\n",
    "\t'POS',\n",
    "\t'REF',\n",
    "\t'ALT',\n",
    "])\n",
    "\n",
    "info_idx = 0,1,3,4  # columns for CHROM, POS, REF, ALT\n",
    "\n",
    "for i, fname in enumerate(dir_list, start=1):\n",
    "\twith open(f\"{EXTRACTED_DIR}/{fname}\", 'r') as f:\n",
    "\t\t# extract DNM information from each line of the file\n",
    "\t\tlines_read = [ [    str.split(line,'\\t')[index] if index else CHROM_DICT[str.split(line,'\\t')[index]] # convert string chromosome ID to integer ID (i.e. 'chrom1' to 1)\n",
    "\t\t\t\t\t\t\tfor index\n",
    "\t\t\t\t\t\t\tin info_idx     ]\n",
    "\t\t\t\t\t\tfor line\n",
    "\t\t\t\t\t\tin f.readlines()\n",
    "\t\t\t\t\t\tif not line[0].startswith('#') ]    # skip lines starting with '#', as they are comments\n",
    "\n",
    "\t\taggregation.extend(lines_read)\n",
    "\t\t\n",
    "\tprint(f\"{i*100/len(dir_list):.0f}% complete\", end='\\r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>619713</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>27343574</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>41444960</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>47323686</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>48889441</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108649</th>\n",
       "      <td>8</td>\n",
       "      <td>140426803</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108650</th>\n",
       "      <td>9</td>\n",
       "      <td>14254838</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108652</th>\n",
       "      <td>9</td>\n",
       "      <td>102594206</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108653</th>\n",
       "      <td>9</td>\n",
       "      <td>113945174</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108654</th>\n",
       "      <td>X</td>\n",
       "      <td>155806128</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101259 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHROM        POS REF ALT\n",
       "0          1     619713   C   T\n",
       "1          1   27343574   G   A\n",
       "2          1   41444960   A   G\n",
       "3          1   47323686   C   T\n",
       "4          1   48889441   A   T\n",
       "...      ...        ...  ..  ..\n",
       "108649     8  140426803   C   G\n",
       "108650     9   14254838   A   T\n",
       "108652     9  102594206   C   T\n",
       "108653     9  113945174   G   A\n",
       "108654     X  155806128   T   C\n",
       "\n",
       "[101259 rows x 4 columns]"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load aggregation into Pandas dataset\n",
    "\n",
    "iceland_columns = aggregation[0] # use first item in aggregation as column headers\n",
    "iceland_dataset = pd.DataFrame(\n",
    "\taggregation[1:],\n",
    "\tcolumns=iceland_columns,\n",
    ")\n",
    "\n",
    "# discard mutations greater than 1 base in length\n",
    "iceland_dataset = iceland_dataset[\n",
    "\t(iceland_dataset.REF.apply(len) == 1) &\n",
    "\t(iceland_dataset.ALT.apply(len) == 1)\n",
    "]\n",
    "iceland_dataset.CHROM = iceland_dataset.CHROM.astype(str)   # ensure POS is integer\n",
    "iceland_dataset.POS = iceland_dataset.POS.astype(int)   # ensure POS is integer\n",
    "\n",
    "iceland_dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write dataset out to file\n",
    "with open(\"./iceland_dataset.csv\", WRITING_MODE) as fout:\n",
    "\ticeland_dataset.to_csv(\n",
    "\t\tfout,\n",
    "\t\tindex=False,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_________"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **2.** <u>Generate windows</u>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This section generates ±2501b windows around the TSS/intergenic window center for each coordinate. The desired window length is ±2.5kb, for a total window size of 5kb, however, 1 extra base is included on either edge of the window. This allows for counting of CpG mutations at the edges."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **2.1** Load hg38 genome\n",
    "\n",
    "This sub-section loads the reference genome into a Biopython dictionary object. The reference genome used in this analysis is the *hg38 genome, release 109*. The FASTA file for this genome can be accessed through the following link:\n",
    "\n",
    "https://ftp.ensembl.org/pub/release-109/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.toplevel.fa.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Please note, `SeqIO.index()` function is slow — this function can take upward of 10 minutes for the FASTA file used in this analysis. Also note, `SeqIO.index()` is very slow to reference items within it as it loads the genome dynamically into memory on demand from disk, which is a very slow process. Use `SeqIO.to_dict()` if you have large RAM (>70GB) for significantly faster indexing, as `SeqIO.to_dict()` loads the entire genome into memory. Both are given in the cell below, but use ONLY one and NOT both.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use this if you have smaller RAM capacity\n",
    "hg38 = SeqIO.index(\n",
    "\tHG38_PATH,\n",
    "\t'fasta',\n",
    ")\n",
    "\n",
    "# # ONLY use this if you have >70GB of RAM\n",
    "# hg38 = SeqIO.to_dict(\n",
    "#     SeqIO.parse(\n",
    "#         HG38_PATH,\n",
    "#         'fasta',\n",
    "#     ),\n",
    "# )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **2.2** Read coordinates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-section reads the coordinates for each coordinate set into a Pandas DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>TSS</th>\n",
       "      <th>FIRST_EIB</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>ENSGID</th>\n",
       "      <th>ENSTID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>923923</td>\n",
       "      <td>924948</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>ENST00000616016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>959256</td>\n",
       "      <td>959215</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>ENST00000327044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>960584</td>\n",
       "      <td>960800</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>ENST00000338591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>966502</td>\n",
       "      <td>966614</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187583</td>\n",
       "      <td>ENST00000379407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1000097</td>\n",
       "      <td>999866</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188290</td>\n",
       "      <td>ENST00000304952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17101</th>\n",
       "      <td>Y</td>\n",
       "      <td>14056227</td>\n",
       "      <td>14056391</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000129862</td>\n",
       "      <td>ENST00000250823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17102</th>\n",
       "      <td>Y</td>\n",
       "      <td>14524529</td>\n",
       "      <td>14524708</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000165246</td>\n",
       "      <td>ENST00000481089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17103</th>\n",
       "      <td>Y</td>\n",
       "      <td>19744726</td>\n",
       "      <td>19744671</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000012817</td>\n",
       "      <td>ENST00000317961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17104</th>\n",
       "      <td>Y</td>\n",
       "      <td>20575776</td>\n",
       "      <td>20575887</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>ENST00000361365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17105</th>\n",
       "      <td>Y</td>\n",
       "      <td>20756108</td>\n",
       "      <td>20756166</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000280969</td>\n",
       "      <td>ENST00000629237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17106 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM       TSS  FIRST_EIB STRAND           ENSGID           ENSTID\n",
       "0         1    923923     924948      +  ENSG00000187634  ENST00000616016\n",
       "1         1    959256     959215      -  ENSG00000188976  ENST00000327044\n",
       "2         1    960584     960800      +  ENSG00000187961  ENST00000338591\n",
       "3         1    966502     966614      +  ENSG00000187583  ENST00000379407\n",
       "4         1   1000097     999866      -  ENSG00000188290  ENST00000304952\n",
       "...     ...       ...        ...    ...              ...              ...\n",
       "17101     Y  14056227   14056391      +  ENSG00000129862  ENST00000250823\n",
       "17102     Y  14524529   14524708      +  ENSG00000165246  ENST00000481089\n",
       "17103     Y  19744726   19744671      -  ENSG00000012817  ENST00000317961\n",
       "17104     Y  20575776   20575887      +  ENSG00000198692  ENST00000361365\n",
       "17105     Y  20756108   20756166      +  ENSG00000280969  ENST00000629237\n",
       "\n",
       "[17106 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_coords_columns = [  # column headers\n",
    "\t\"CHROM\",\n",
    "\t\"TSS\",\n",
    "\t\"FIRST_EIB\",\n",
    "\t\"STRAND\",\n",
    "\t\"ENSGID\",\n",
    "\t\"ENSTID\",\n",
    "]\n",
    "\n",
    "tss_coords = pd.read_csv(\n",
    "\tTSS_COORDS_PATH,\n",
    "\tdelimiter='\\t',\n",
    "\tskiprows=[0],\n",
    "\tnames=tss_coords_columns,\n",
    ")\n",
    "tss_coords.replace({'CHROM':CHROM_DICT}, inplace=True)  # chromosome string formatting (\"chrom#\" -> # ; i.e. 'chrome1' -> 1)\n",
    "tss_coords = tss_coords[tss_coords.CHROM != 'M']        # drop mitochondrial DNA DNMs \n",
    "tss_coords.TSS = tss_coords.TSS.astype(int)             # ensure TSS is integer\n",
    "tss_coords.FIRST_EIB = tss_coords.FIRST_EIB.astype(int) # ensure FIRST_EIB is integer\n",
    "tss_coords = tss_coords.drop_duplicates()\n",
    "\n",
    "tss_coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>114327737</td>\n",
       "      <td>114327837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>114328137</td>\n",
       "      <td>114328237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>114328237</td>\n",
       "      <td>114328337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>47396304</td>\n",
       "      <td>47396404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>166420304</td>\n",
       "      <td>166420404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17107</th>\n",
       "      <td>21</td>\n",
       "      <td>8439401</td>\n",
       "      <td>8439501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17108</th>\n",
       "      <td>22</td>\n",
       "      <td>44331786</td>\n",
       "      <td>44331886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17109</th>\n",
       "      <td>22</td>\n",
       "      <td>46071211</td>\n",
       "      <td>46071311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17110</th>\n",
       "      <td>22</td>\n",
       "      <td>16601785</td>\n",
       "      <td>16601885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17111</th>\n",
       "      <td>22</td>\n",
       "      <td>48086859</td>\n",
       "      <td>48086959</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16616 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM      START        END\n",
       "0         1  114327737  114327837\n",
       "1         1  114328137  114328237\n",
       "2         1  114328237  114328337\n",
       "3         1   47396304   47396404\n",
       "4         1  166420304  166420404\n",
       "...     ...        ...        ...\n",
       "17107    21    8439401    8439501\n",
       "17108    22   44331786   44331886\n",
       "17109    22   46071211   46071311\n",
       "17110    22   16601785   16601885\n",
       "17111    22   48086859   48086959\n",
       "\n",
       "[16616 rows x 3 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergen_coords_columns = [ # column headers\n",
    "\t\"CHROM\",\n",
    "\t\"START\",\n",
    "\t\"END\",\n",
    "]\n",
    "\n",
    "intergen_GCmatch_coords = pd.read_csv(\n",
    "\tINTERGEN_GCMATCH_COORDS_PATH,\n",
    "\tdelimiter='\\t',\n",
    "\tskiprows=[0],\n",
    "\tnames=intergen_coords_columns,\n",
    ")\n",
    "intergen_GCmatch_coords.replace({'CHROM':CHROM_DICT}, inplace=True)         # chromosome string formatting (\"chrom#\" -> # ; i.e. 'chrome1' -> 1)\n",
    "intergen_GCmatch_coords.CHROM = intergen_GCmatch_coords.CHROM.astype(str)   # ensure CHROM is string\n",
    "intergen_GCmatch_coords.START = intergen_GCmatch_coords.START.astype(int)   # ensure START is integer\n",
    "intergen_GCmatch_coords.END = intergen_GCmatch_coords.END.astype(int)       # ensure END is integer\n",
    "intergen_GCmatch_coords = intergen_GCmatch_coords.drop_duplicates()\n",
    "\n",
    "intergen_GCmatch_coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>128802631</td>\n",
       "      <td>128807631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>139115639</td>\n",
       "      <td>139120639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>114326537</td>\n",
       "      <td>114331537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>47396104</td>\n",
       "      <td>47401104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>166419304</td>\n",
       "      <td>166424304</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177995</th>\n",
       "      <td>22</td>\n",
       "      <td>29301107</td>\n",
       "      <td>29306107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177996</th>\n",
       "      <td>22</td>\n",
       "      <td>6327835</td>\n",
       "      <td>6332835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177997</th>\n",
       "      <td>22</td>\n",
       "      <td>470324</td>\n",
       "      <td>475324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177998</th>\n",
       "      <td>22</td>\n",
       "      <td>11498452</td>\n",
       "      <td>11503452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177999</th>\n",
       "      <td>22</td>\n",
       "      <td>30630117</td>\n",
       "      <td>30635117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>178000 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHROM      START        END\n",
       "0           1  128802631  128807631\n",
       "1           1  139115639  139120639\n",
       "2           1  114326537  114331537\n",
       "3           1   47396104   47401104\n",
       "4           1  166419304  166424304\n",
       "...       ...        ...        ...\n",
       "177995     22   29301107   29306107\n",
       "177996     22    6327835    6332835\n",
       "177997     22     470324     475324\n",
       "177998     22   11498452   11503452\n",
       "177999     22   30630117   30635117\n",
       "\n",
       "[178000 rows x 3 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergen_nonGCmatch_coords = pd.read_csv(\n",
    "\tINTERGEN_NO_GCMATCH_COORDS_PATH,\n",
    "\tdelimiter='\\t',\n",
    "\tskiprows=[0],\n",
    "\tnames=intergen_coords_columns,\n",
    ")\n",
    "intergen_nonGCmatch_coords.replace({'CHROM':CHROM_DICT}, inplace=True)          # chromosome string formatting (\"chrom#\" -> # ; i.e. 'chrome1' -> 1)\n",
    "intergen_nonGCmatch_coords.CHROM = intergen_nonGCmatch_coords.CHROM.astype(str) # ensure CHROM is string\n",
    "intergen_nonGCmatch_coords.START = intergen_nonGCmatch_coords.START.astype(int) # ensure START is integer\n",
    "intergen_nonGCmatch_coords.END = intergen_nonGCmatch_coords.END.astype(int)     # ensure END is integer\n",
    "intergen_nonGCmatch_coords = intergen_nonGCmatch_coords.drop_duplicates()\n",
    "\n",
    "intergen_nonGCmatch_coords"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **2.3** Generate windows"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-section generates a string representation of the ±2501 base window centered around each coordinate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop coordinates with windows containing bases other than 'A', 'C', 'G', 'T'\n",
    "\n",
    "def drop_invalid_windows(windows:pd.DataFrame, valid_bases:set[str], window_col:int=-1, print_dropped:bool=True) -> None:\n",
    "\t'''\n",
    "\tDrop coordinates inplace with windows containing bases other than those specified in `valid_bases`.\n",
    "\n",
    "\t## Parameters\n",
    "\twindows: DataFrame\n",
    "\t\tDataFrame containing windows\n",
    "\tvalid_bases: set\n",
    "\t\tSet of strings representing valid bases\n",
    "\twindow_col: int, default = -1\n",
    "\t\tColumn index in `windows` containing the window string\n",
    "\tprint_dropped: bool, default = True\n",
    "\t\tIf True, indices of dropped windows, along with a set containing\n",
    "\t\tthe invalid bases found in the window, are printed\n",
    "\t'''\n",
    "\n",
    "\tany_dropped = False\n",
    "\n",
    "\tfor i, coord in enumerate(windows.values):\n",
    "\t\twindow = coord[window_col]\n",
    "\n",
    "\t\tif not set(window) <= valid_bases:\n",
    "\t\t\twindows.drop(i, inplace=True)\n",
    "\t\t\t\n",
    "\t\t\tif print_dropped:\n",
    "\t\t\t\tif not any_dropped:\n",
    "\t\t\t\t\tprint(\"Dropping\", end=' ')\n",
    "\t\t\t\t\tany_dropped = True\n",
    "\t\t\t\tprint(f\"{i} (contains invalid base {set(window) - valid_bases})\", end=', ')\n",
    "\n",
    "\tif not any_dropped:\n",
    "\t\tprint(\"No windows dropped\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **2.3.1** TSS windows"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this sub-sub-section, the TSS is used as the center."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This method packs the data together in a list (`tss_windows`) of lists. Each sublist contains the following information for each coordinate: the chromosome, tss, first eib, stradedness, ensGID, ensTID, and 5kb window.\n",
    "\n",
    "Different offsets for positive and negative strands are accounted for. Positive strands are indexed as [-2501, 2500) around the TSS. Negative strands are indexed as [-2500, 2501) around the TSS, and are reversed. This keeps the TSS at index 0 when interpreting the range as [-2501,2500]. The left half of the window is the intergenic space immediately before the gene, and the right half is the gene itself."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>TSS</th>\n",
       "      <th>FIRST_EIB</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>ENSGID</th>\n",
       "      <th>ENSTID</th>\n",
       "      <th>WINDOW</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>923923</td>\n",
       "      <td>924948</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187634</td>\n",
       "      <td>ENST00000616016</td>\n",
       "      <td>GGGAGAGTGGGGCCCAGGTGTCCCAGACCAAGACAGGTGTGTGAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>959256</td>\n",
       "      <td>959215</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188976</td>\n",
       "      <td>ENST00000327044</td>\n",
       "      <td>GGGAGTGGACGTGTAACGGGAGCGGGTGGTGTTAGAGTCGGCACAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>960584</td>\n",
       "      <td>960800</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187961</td>\n",
       "      <td>ENST00000338591</td>\n",
       "      <td>TCCTGAGGCGGAATCTCACTCTGTCGCCCAGGCTGGAGTGCAGTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>966502</td>\n",
       "      <td>966614</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187583</td>\n",
       "      <td>ENST00000379407</td>\n",
       "      <td>CACGCTTGGTGGGTGATGGGGCCTGCCTGGGGGGCATCCCCACCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1000097</td>\n",
       "      <td>999866</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188290</td>\n",
       "      <td>ENST00000304952</td>\n",
       "      <td>GGGGTAGAGATGATTTTTATGTTTTGTGATCGGCCCGCACCACCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17101</th>\n",
       "      <td>Y</td>\n",
       "      <td>14056227</td>\n",
       "      <td>14056391</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000129862</td>\n",
       "      <td>ENST00000250823</td>\n",
       "      <td>TAAAAATAAAATCACATTCTTCTCACCACGTCCTGCTGAGTTATTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17102</th>\n",
       "      <td>Y</td>\n",
       "      <td>14524529</td>\n",
       "      <td>14524708</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000165246</td>\n",
       "      <td>ENST00000481089</td>\n",
       "      <td>GAGTATTTAGAAACAACTGCCTCTTGGTAAGCACACCTCCTATATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17103</th>\n",
       "      <td>Y</td>\n",
       "      <td>19744726</td>\n",
       "      <td>19744671</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000012817</td>\n",
       "      <td>ENST00000317961</td>\n",
       "      <td>ACAGAGTCAGGGTCGAACGAACAGACAACTTTGGGGTCGAACAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17104</th>\n",
       "      <td>Y</td>\n",
       "      <td>20575776</td>\n",
       "      <td>20575887</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>ENST00000361365</td>\n",
       "      <td>CTGTCTTCTTCTAGGCCCTTCAAACTGTTCCAACCTCCCCCCATTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17105</th>\n",
       "      <td>Y</td>\n",
       "      <td>20756108</td>\n",
       "      <td>20756166</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000280969</td>\n",
       "      <td>ENST00000629237</td>\n",
       "      <td>GTATTTGACTAACAGTCTTCTGAAAATGATTCAATCAGAAAAGCCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17106 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM       TSS  FIRST_EIB STRAND           ENSGID           ENSTID  \\\n",
       "0         1    923923     924948      +  ENSG00000187634  ENST00000616016   \n",
       "1         1    959256     959215      -  ENSG00000188976  ENST00000327044   \n",
       "2         1    960584     960800      +  ENSG00000187961  ENST00000338591   \n",
       "3         1    966502     966614      +  ENSG00000187583  ENST00000379407   \n",
       "4         1   1000097     999866      -  ENSG00000188290  ENST00000304952   \n",
       "...     ...       ...        ...    ...              ...              ...   \n",
       "17101     Y  14056227   14056391      +  ENSG00000129862  ENST00000250823   \n",
       "17102     Y  14524529   14524708      +  ENSG00000165246  ENST00000481089   \n",
       "17103     Y  19744726   19744671      -  ENSG00000012817  ENST00000317961   \n",
       "17104     Y  20575776   20575887      +  ENSG00000198692  ENST00000361365   \n",
       "17105     Y  20756108   20756166      +  ENSG00000280969  ENST00000629237   \n",
       "\n",
       "                                                  WINDOW  \n",
       "0      GGGAGAGTGGGGCCCAGGTGTCCCAGACCAAGACAGGTGTGTGAGT...  \n",
       "1      GGGAGTGGACGTGTAACGGGAGCGGGTGGTGTTAGAGTCGGCACAT...  \n",
       "2      TCCTGAGGCGGAATCTCACTCTGTCGCCCAGGCTGGAGTGCAGTGG...  \n",
       "3      CACGCTTGGTGGGTGATGGGGCCTGCCTGGGGGGCATCCCCACCTT...  \n",
       "4      GGGGTAGAGATGATTTTTATGTTTTGTGATCGGCCCGCACCACCAT...  \n",
       "...                                                  ...  \n",
       "17101  TAAAAATAAAATCACATTCTTCTCACCACGTCCTGCTGAGTTATTG...  \n",
       "17102  GAGTATTTAGAAACAACTGCCTCTTGGTAAGCACACCTCCTATATT...  \n",
       "17103  ACAGAGTCAGGGTCGAACGAACAGACAACTTTGGGGTCGAACAAAG...  \n",
       "17104  CTGTCTTCTTCTAGGCCCTTCAAACTGTTCCAACCTCCCCCCATTA...  \n",
       "17105  GTATTTGACTAACAGTCTTCTGAAAATGATTCAATCAGAAAAGCCC...  \n",
       "\n",
       "[17106 rows x 7 columns]"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_windows = list()\n",
    "for chrom, tss, first_eib, strand, ensgid, enstid in tss_coords.values:\n",
    "\tassert strand in '+-'\n",
    "\n",
    "\tchrom_value = hg38[str(chrom)]\n",
    "\tif strand == '+':\n",
    "\t\tstart_idx = tss - (WINDOW_LENGTH//2 + 1) - CPG_OFFSET\n",
    "\t\tstop_idx = tss + (WINDOW_LENGTH//2 - 1) + CPG_OFFSET\n",
    "\t\tdirection = 1\n",
    "\n",
    "\telif strand == '-':\n",
    "\t\tstart_idx = tss - WINDOW_LENGTH//2 - CPG_OFFSET\n",
    "\t\tstop_idx = tss + WINDOW_LENGTH//2 + CPG_OFFSET\n",
    "\t\tdirection = -1\n",
    "\n",
    "\n",
    "\ttss_windows.append(\n",
    "\t\t[\n",
    "\t\t\tchrom,\n",
    "\t\t\ttss,\n",
    "\t\t\tfirst_eib,\n",
    "\t\t\tstrand,\n",
    "\t\t\tensgid,\n",
    "\t\t\tenstid,\n",
    "\t\t\tstr(chrom_value[start_idx : stop_idx].seq)[::direction],   # string representation of ±2501 base window\n",
    "\t\t]\n",
    "\t)\n",
    "\n",
    "\n",
    "columns = [\t# column headers\n",
    "\t'CHROM',\n",
    "\t'TSS',\n",
    "\t'FIRST_EIB',\n",
    "\t'STRAND',\n",
    "\t'ENSGID',\n",
    "\t'ENSTID',\n",
    "\t'WINDOW',\n",
    "]\n",
    "tss_windows = pd.DataFrame(\n",
    "\ttss_windows,\n",
    "\tcolumns=columns,\n",
    ")\n",
    "\n",
    "# explicit casting of column types to ensure consistent results\n",
    "tss_windows.CHROM = tss_windows.CHROM.astype(str)\n",
    "tss_windows.TSS = tss_windows.TSS.astype(int)\n",
    "tss_windows.FIRST_EIB = tss_windows.FIRST_EIB.astype(int)\n",
    "\n",
    "tss_windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropping 8440 (contains invalid base {'N'}), 8441 (contains invalid base {'N'}), "
     ]
    }
   ],
   "source": [
    "# drop coordinates with windows containing bases other than 'A', 'C', 'G', 'T'\n",
    "\n",
    "drop_invalid_windows(\n",
    "\twindows=tss_windows,\n",
    "\tvalid_bases=BASES_SET,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{WINDOWS_DIR}/hg38_tss.csv\", WRITING_MODE) as fout:\n",
    "\ttss_windows.to_csv(\n",
    "\t\tfout,\n",
    "\t\tindex=False,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **2.3.2** Intergenic windows"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this sub-sub-section, the intergenic coordinate is used as the center, for both the GC matched and non-GC matched coordinates. The same method used in sub-sub-section 2.3.1 for the TSS coordinates is applied here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_intergen_windows(coords: pd.DataFrame, genome: dict, win_len: int, offset: int=0) -> list:\n",
    "\t'''\n",
    "\tGenerate windows at center of coordinates specified in `coords`.\n",
    "\n",
    "\t## Parameters\n",
    "\tcoords: Pandas.DataFrame\n",
    "\t\tPandas.DataFrame containing coordinates.\n",
    "\tgenome: dict\n",
    "\t\tDictionary returned by Biopython's SeqIO interface. Contains the genome to \n",
    "\t\tbe used to generate the windows.\n",
    "\twin_len: int\n",
    "\t\tLength of window.\n",
    "\toffset: int, default = 0\n",
    "\t\tWindow offset.\n",
    "\n",
    "\t## Returns\n",
    "\twindows: list\n",
    "\t\tList with each item containing the following information for each coordinate: \n",
    "\t\t\n",
    "\t\t\t- chromosome\n",
    "\t\t\t- coordinate for center of window\n",
    "\t\t\t- window\n",
    "\t'''\n",
    "\n",
    "\twindows = list()\n",
    "\tfor chrom, start, end in coords.values:\n",
    "\t\tchrom_value = genome[str(chrom)]\n",
    "\n",
    "\t\tstart_idx = start + (end - start - win_len)//2 - offset\n",
    "\t\tstop_idx = start + (end - start + win_len)//2 + offset\n",
    "\n",
    "\t\twindows.append(\n",
    "\t\t\t[\n",
    "\t\t\t\tchrom,\n",
    "\t\t\t\tstart + (end - start)//2,\n",
    "\t\t\t\tstr(chrom_value[start_idx : stop_idx].seq),   # string representation of ±2501 base window\n",
    "\t\t\t]\n",
    "\t\t)\n",
    "\n",
    "\treturn windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergen_windows_columns = [\t# column headers\n",
    "\t'CHROM',\n",
    "\t'CENTER',\n",
    "\t'WINDOW',\n",
    "]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### *GC matched*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>CENTER</th>\n",
       "      <th>WINDOW</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>114327787</td>\n",
       "      <td>GATAGCAGCAGTTCTTATTTTTTGAAGCCAATATTACCTTGATATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>114328187</td>\n",
       "      <td>ACATCATACTCAGTAGTGAAGGACTAAATGCTTTCTCAAATACCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>114328287</td>\n",
       "      <td>GAACAATAAAGAAAGAAAAATAATTAACAGGCTTATAGATTGGAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>47396354</td>\n",
       "      <td>TTGCTTTTTGCACAATGGTAACACATATGGACTCCCTCTTTTTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>166420354</td>\n",
       "      <td>GTGTTGAAAGATAGCACTCATCCTGATGAGTTAAACAGTTAAAGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16611</th>\n",
       "      <td>21</td>\n",
       "      <td>8439451</td>\n",
       "      <td>TACAGTGAAACTGCGAATGGCTCATTAAATCAGTTATGGTTCCTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16612</th>\n",
       "      <td>22</td>\n",
       "      <td>44331836</td>\n",
       "      <td>TCACAGGTCTGAGTGAGCAGCACTGGCCCTGCCCCCTACCCCTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16613</th>\n",
       "      <td>22</td>\n",
       "      <td>46071261</td>\n",
       "      <td>ATTGCCCTTGAGAACAGTTTTATTGAGAATAGTAACCATATATTTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16614</th>\n",
       "      <td>22</td>\n",
       "      <td>16601835</td>\n",
       "      <td>AATTTAAAATCCCATATTTAACATGTATACTTATTAGAAAGTACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16615</th>\n",
       "      <td>22</td>\n",
       "      <td>48086909</td>\n",
       "      <td>GATGGAAACCACACACAAATGTCAAGTGGCCGTCAACAACCGGGGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16616 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM     CENTER                                             WINDOW\n",
       "0         1  114327787  GATAGCAGCAGTTCTTATTTTTTGAAGCCAATATTACCTTGATATA...\n",
       "1         1  114328187  ACATCATACTCAGTAGTGAAGGACTAAATGCTTTCTCAAATACCTG...\n",
       "2         1  114328287  GAACAATAAAGAAAGAAAAATAATTAACAGGCTTATAGATTGGAAA...\n",
       "3         1   47396354  TTGCTTTTTGCACAATGGTAACACATATGGACTCCCTCTTTTTTTT...\n",
       "4         1  166420354  GTGTTGAAAGATAGCACTCATCCTGATGAGTTAAACAGTTAAAGGT...\n",
       "...     ...        ...                                                ...\n",
       "16611    21    8439451  TACAGTGAAACTGCGAATGGCTCATTAAATCAGTTATGGTTCCTTT...\n",
       "16612    22   44331836  TCACAGGTCTGAGTGAGCAGCACTGGCCCTGCCCCCTACCCCTCTC...\n",
       "16613    22   46071261  ATTGCCCTTGAGAACAGTTTTATTGAGAATAGTAACCATATATTTA...\n",
       "16614    22   16601835  AATTTAAAATCCCATATTTAACATGTATACTTATTAGAAAGTACAC...\n",
       "16615    22   48086909  GATGGAAACCACACACAAATGTCAAGTGGCCGTCAACAACCGGGGC...\n",
       "\n",
       "[16616 rows x 3 columns]"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergen_GCmatch_windows = get_intergen_windows(\n",
    "    coords=intergen_GCmatch_coords,\n",
    "    genome=hg38,\n",
    "    win_len=WINDOW_LENGTH,\n",
    "\toffset=CPG_OFFSET,\n",
    ")\n",
    "intergen_GCmatch_windows = pd.DataFrame(\n",
    "\tintergen_GCmatch_windows,\n",
    "\tcolumns=intergen_windows_columns,\n",
    ")\n",
    "\n",
    "# explicit casting of column types to ensure consistent results\n",
    "intergen_GCmatch_windows.CHROM = intergen_GCmatch_windows.CHROM.astype(str)\n",
    "intergen_GCmatch_windows.CENTER = intergen_GCmatch_windows.CENTER.astype(int)\n",
    "\n",
    "intergen_GCmatch_windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dropping 5084 (contains invalid base {'N'}), 5085 (contains invalid base {'N'}), 5951 (contains invalid base {'N'}), 8565 (contains invalid base {'N'}), 8566 (contains invalid base {'N'}), 8567 (contains invalid base {'N'}), 8568 (contains invalid base {'N'}), 8569 (contains invalid base {'N'}), 8570 (contains invalid base {'N'}), 9881 (contains invalid base {'N'}), 9882 (contains invalid base {'N'}), 9883 (contains invalid base {'N'}), 10341 (contains invalid base {'N'}), 10810 (contains invalid base {'N'}), 10811 (contains invalid base {'N'}), 10878 (contains invalid base {'N'}), 10879 (contains invalid base {'N'}), 12246 (contains invalid base {'N'}), 13132 (contains invalid base {'N'}), 13133 (contains invalid base {'N'}), 13134 (contains invalid base {'N'}), 13135 (contains invalid base {'N'}), 13136 (contains invalid base {'N'}), 13137 (contains invalid base {'N'}), 13548 (contains invalid base {'N'}), 14915 (contains invalid base {'N'}), 15010 (contains invalid base {'N'}), 15011 (contains invalid base {'N'}), 15012 (contains invalid base {'N'}), 15025 (contains invalid base {'N'}), 15130 (contains invalid base {'N'}), 15210 (contains invalid base {'N'}), 15211 (contains invalid base {'N'}), 15212 (contains invalid base {'N'}), 15213 (contains invalid base {'N'}), 15931 (contains invalid base {'N'}), 15932 (contains invalid base {'N'}), 15933 (contains invalid base {'N'}), 15947 (contains invalid base {'N'}), 16169 (contains invalid base {'N'}), 16170 (contains invalid base {'N'}), 16412 (contains invalid base {'N'}), 16413 (contains invalid base {'N'}), 16582 (contains invalid base {'N'}), 16583 (contains invalid base {'N'}), 16584 (contains invalid base {'N'}), 16600 (contains invalid base {'N'}), 16610 (contains invalid base {'N'}), "
     ]
    }
   ],
   "source": [
    "# drop coordinates with windows containing bases other than 'A', 'C', 'G', 'T'\n",
    "\n",
    "drop_invalid_windows(\n",
    "\twindows=intergen_GCmatch_windows,\n",
    "\tvalid_bases=BASES_SET,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{WINDOWS_DIR}/hg38_intergen_GCmatch.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_GCmatch_windows.to_csv(\n",
    "\t\tfout,\n",
    "\t\tindex=False,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### *non-GC matched*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>CENTER</th>\n",
       "      <th>WINDOW</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>114329037</td>\n",
       "      <td>TGAGATTTATATAATCTTGTACTCTCAGCACTTAGAGCAGAACAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>47398604</td>\n",
       "      <td>AGGACTACCGGTGTGCACCACCATGCCAGGCTAATTTTTGTATTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>166421804</td>\n",
       "      <td>GAGGGATTGTCAACATGAAAACCACCATGAAGAGGTAAGAATGGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>69215104</td>\n",
       "      <td>CAAAAGAAGGGGGGAGGTCTCTTCCTTTGTTTGCAATAGGGAGTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>29412419</td>\n",
       "      <td>CTGGACAGAAAAAGTGGAGTTTATGTGCAAAATTTATGTGCAAAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177991</th>\n",
       "      <td>22</td>\n",
       "      <td>11487044</td>\n",
       "      <td>CTTCCCAAAGGTGGAAGCAGGAAGCGAGTTTTCTCAGGTGTTTGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177993</th>\n",
       "      <td>22</td>\n",
       "      <td>27193525</td>\n",
       "      <td>TGACTTAAATAGTGAAGAGGAGACAGTTTGCTTCAATGCTTGGCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177994</th>\n",
       "      <td>22</td>\n",
       "      <td>12064284</td>\n",
       "      <td>TAACTAAAGAAAATTAAGATAAAATAATTTTGACAACAGCTTGACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177995</th>\n",
       "      <td>22</td>\n",
       "      <td>29303607</td>\n",
       "      <td>CTCTCAGTCACCAACCACCTCTAATGTTGCCCAATCAATGGTCTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177999</th>\n",
       "      <td>22</td>\n",
       "      <td>30632617</td>\n",
       "      <td>CCCCTCTGGGCTCCCATGGAGGATGGTGCTGCTGGGTCACCTGCTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163508 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       CHROM     CENTER                                             WINDOW\n",
       "2          1  114329037  TGAGATTTATATAATCTTGTACTCTCAGCACTTAGAGCAGAACAAG...\n",
       "3          1   47398604  AGGACTACCGGTGTGCACCACCATGCCAGGCTAATTTTTGTATTTT...\n",
       "4          1  166421804  GAGGGATTGTCAACATGAAAACCACCATGAAGAGGTAAGAATGGTT...\n",
       "5          1   69215104  CAAAAGAAGGGGGGAGGTCTCTTCCTTTGTTTGCAATAGGGAGTAT...\n",
       "6          1   29412419  CTGGACAGAAAAAGTGGAGTTTATGTGCAAAATTTATGTGCAAAGG...\n",
       "...      ...        ...                                                ...\n",
       "177991    22   11487044  CTTCCCAAAGGTGGAAGCAGGAAGCGAGTTTTCTCAGGTGTTTGCC...\n",
       "177993    22   27193525  TGACTTAAATAGTGAAGAGGAGACAGTTTGCTTCAATGCTTGGCAC...\n",
       "177994    22   12064284  TAACTAAAGAAAATTAAGATAAAATAATTTTGACAACAGCTTGACC...\n",
       "177995    22   29303607  CTCTCAGTCACCAACCACCTCTAATGTTGCCCAATCAATGGTCTTT...\n",
       "177999    22   30632617  CCCCTCTGGGCTCCCATGGAGGATGGTGCTGCTGGGTCACCTGCTT...\n",
       "\n",
       "[163508 rows x 3 columns]"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergen_nonGCmatch_windows = get_intergen_windows(\n",
    "    coords=intergen_nonGCmatch_coords,\n",
    "    genome=hg38,\n",
    "    win_len=WINDOW_LENGTH,\n",
    "\toffset=CPG_OFFSET,\n",
    ")\n",
    "intergen_nonGCmatch_windows = pd.DataFrame(\n",
    "\tintergen_nonGCmatch_windows,\n",
    "\tcolumns=intergen_windows_columns,\n",
    ")\n",
    "intergen_nonGCmatch_windows = intergen_nonGCmatch_windows[~intergen_nonGCmatch_windows.WINDOW.str.contains('N')]\t# explicitly drop rows with windows containing 'N' for ease of filtering for other illegal bases\n",
    "\n",
    "# explicit casting of column types to ensure consistent results\n",
    "intergen_nonGCmatch_windows.CHROM = intergen_nonGCmatch_windows.CHROM.astype(str)\n",
    "intergen_nonGCmatch_windows.CENTER = intergen_nonGCmatch_windows.CENTER.astype(int)\n",
    "\n",
    "intergen_nonGCmatch_windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No windows dropped\n"
     ]
    }
   ],
   "source": [
    "# drop coordinates with windows containing bases other than 'A', 'C', 'G', 'T'\n",
    "\n",
    "drop_invalid_windows(\n",
    "\twindows=intergen_nonGCmatch_windows,\n",
    "\tvalid_bases=BASES_SET,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 826,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{WINDOWS_DIR}/hg38_intergen_nonGCmatch.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_nonGCmatch_windows.to_csv(\n",
    "\t\tfout,\n",
    "\t\tindex=False,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **3.** <u>Generate matches</u>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This section finds DNMs that fall into the windows (\"matches\"). This is done for all 3 sets of coordinates (TSS regions, GC matched intergenic, and non-GC matched intergenic)."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **3.1** TSS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-section finds matches for the TSS coordinates. Matches are aggregatedc in a list (`tss_matches_list`) and subsequently loaded into a Pandas DataFrame (`tss_matches`).\n",
    "\n",
    "Each match is packed with the following metrics:\n",
    "\n",
    "|Metric|Description|\n",
    "|-:|-|\n",
    "|`CHROM`|chromosome of the DNM|\n",
    "|`POS`|DNM position on chromosome CHROM|\n",
    "|`REF`|base at position POS in reference genome|\n",
    "|`ALT`|base found at position POS in proband|\n",
    "|`TSS`|transcriptional start site coordinate|\n",
    "|`FIRST_EIB`|coordinate of first extron-intron boundary|\n",
    "|`DIST`|distance from TSS|\n",
    "|`FIRST_EX`|*boolean*: within first exon?|\n",
    "|`DOWNSTREAM`|*boolean*: downstream and after first exon?|\n",
    "|`INTERGENIC`|*boolean*: within intergenic space immediately before TSS?|\n",
    "|`STRAND`|strandedness|\n",
    "|`ENSGID`|Ensebl gene ID|\n",
    "|`ENSTID`|Ensebl transcript ID|\n",
    "|`WINDOW`|string representing ±2501 base window centered around TSS|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1%\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(tss_matches_list) = 2950\n"
     ]
    }
   ],
   "source": [
    "POS_INDEX = 1   # column index for DNM position in dataset DataFrame\n",
    "\n",
    "tss_matches_list = list()\n",
    "for i, tss_coord in enumerate(tss_windows.values, start=1):\n",
    "\tchrom, tss, eib, strandedness, ensGID, ensTID, window = tss_coord\n",
    "\t\n",
    "\t# calculate window start and stop indices\n",
    "\ttss_window_start, tss_window_stop = tss - WINDOW_LENGTH//2, tss + WINDOW_LENGTH//2\n",
    "\n",
    "\t# filter for matches\n",
    "\tmatches_df_temp = iceland_dataset[\n",
    "\t\t(iceland_dataset.POS >= tss_window_start) &\n",
    "\t\t(iceland_dataset.POS <  tss_window_stop) &\n",
    "\t\t(iceland_dataset.CHROM == chrom)\n",
    "\t]\n",
    "\t\n",
    "\tif len(matches_df_temp):\n",
    "\t\t# if matches exist, gather metrics for each\n",
    "\n",
    "\t\tmatch_data_temp = list()\n",
    "\t\tfor item in matches_df_temp.values:\n",
    "\t\t\tif strandedness == '+':\n",
    "\t\t\t\tdist = item[POS_INDEX] - tss\n",
    "\t\t\t\tfirst_ex = int(item[POS_INDEX] <= eib & item[POS_INDEX] >= tss)\n",
    "\t\t\t\tdownstream = int(item[POS_INDEX] >= eib)\n",
    "\t\t\t\tintergenic = int(item[POS_INDEX] < tss)\n",
    "\n",
    "\t\t\telif strandedness == '-':\n",
    "\t\t\t\tdist = tss - item[POS_INDEX]\n",
    "\t\t\t\tfirst_ex = int(item[POS_INDEX] <= tss & item[POS_INDEX] >= eib)\n",
    "\t\t\t\tdownstream = int(item[POS_INDEX] <= eib)\n",
    "\t\t\t\tintergenic = int(item[POS_INDEX] > tss)\n",
    "\n",
    "\t\t\tmatch_data_temp.append(\n",
    "\t\t\t\t[\n",
    "\t\t\t\t\t*item,\t# contains CHROM, POS, REF, ALT\n",
    "\t\t\t\t\ttss,\n",
    "\t\t\t\t\teib,\n",
    "\t\t\t\t\tdist,\n",
    "\t\t\t\t\tfirst_ex,\n",
    "\t\t\t\t\tdownstream,\n",
    "\t\t\t\t\tintergenic,\n",
    "\t\t\t\t\tstrandedness, \n",
    "\t\t\t\t\tensGID,\n",
    "\t\t\t\t\tensTID,\n",
    "\t\t\t\t\twindow,\n",
    "\t\t\t\t]\n",
    "\t\t\t)\n",
    "\n",
    "\t\ttss_matches_list.extend(match_data_temp)\n",
    "\n",
    "\tprint(f\"{i * 100 / len(tss_coords):.0f}%\", end='\\r')\n",
    "\n",
    "print(f\"{len(tss_matches_list) = }\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>TSS</th>\n",
       "      <th>FIRST_EIB</th>\n",
       "      <th>DIST</th>\n",
       "      <th>FIRST_EX</th>\n",
       "      <th>DOWNSTREAM</th>\n",
       "      <th>INTERGENIC</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>ENSGID</th>\n",
       "      <th>ENSTID</th>\n",
       "      <th>WINDOW</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1014727</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>1013497</td>\n",
       "      <td>1013576</td>\n",
       "      <td>1230</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187608</td>\n",
       "      <td>ENST00000649529</td>\n",
       "      <td>CCTCGGTCTCTCAAAGTGCTGGGATTACAGGTGAGCCACTGCGCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1021242</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1020120</td>\n",
       "      <td>1020373</td>\n",
       "      <td>1122</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>ENST00000379370</td>\n",
       "      <td>ATCTGGCCCCACCCACATCCTGCTGATGGGTAGAGCCTAGTGGTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1208352</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1206571</td>\n",
       "      <td>1206385</td>\n",
       "      <td>-1781</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000186891</td>\n",
       "      <td>ENST00000328596</td>\n",
       "      <td>GGGGGCGTCGTATCCCTTCAGTCTCCGGCAGGACAGGGGGTAGGTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1360908</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1358543</td>\n",
       "      <td>1358456</td>\n",
       "      <td>-2365</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000162576</td>\n",
       "      <td>ENST00000445648</td>\n",
       "      <td>ACAGAAGTACACGTACGTACACATACACAGACACAGAAGCGCACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1398713</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1399311</td>\n",
       "      <td>1399019</td>\n",
       "      <td>598</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000221978</td>\n",
       "      <td>ENST00000481223</td>\n",
       "      <td>GGTGGGACCGAGGGGAGTTCCCGACGCACGTGTCACGGGCGAAAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2945</th>\n",
       "      <td>X</td>\n",
       "      <td>151976628</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>151974676</td>\n",
       "      <td>151974570</td>\n",
       "      <td>-1952</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000102287</td>\n",
       "      <td>ENST00000370328</td>\n",
       "      <td>GAGACAGTTTGTGGTGTGTATCGACCAAAGTTGTGGGACCGACCGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2946</th>\n",
       "      <td>X</td>\n",
       "      <td>152735883</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>152733735</td>\n",
       "      <td>152733668</td>\n",
       "      <td>-2148</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000198930</td>\n",
       "      <td>ENST00000370287</td>\n",
       "      <td>GTTCGGGTCCCGGAGAGGAACCCGGAGTTCCGGAAGGAGTCCGAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2947</th>\n",
       "      <td>X</td>\n",
       "      <td>152735883</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>152733757</td>\n",
       "      <td>152733859</td>\n",
       "      <td>2126</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000213401</td>\n",
       "      <td>ENST00000393869</td>\n",
       "      <td>TAATAAAACGTTTTTTATCACAAATATGCCATTAAAAGTGAAATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2948</th>\n",
       "      <td>X</td>\n",
       "      <td>153727239</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>153724856</td>\n",
       "      <td>153726166</td>\n",
       "      <td>2383</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000101986</td>\n",
       "      <td>ENST00000218104</td>\n",
       "      <td>TTCATTTGCTTTTCCAGTTTCCTTTGTGTAGAAATCAGATTGTCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2949</th>\n",
       "      <td>X</td>\n",
       "      <td>154749911</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>154751077</td>\n",
       "      <td>154750954</td>\n",
       "      <td>1166</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000160219</td>\n",
       "      <td>ENST00000424127</td>\n",
       "      <td>AGGAAACGGTTTACAATAAAAAAGGGTAAGTCGTCTCTTGTGATCT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2950 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROM        POS REF ALT        TSS  FIRST_EIB  DIST  FIRST_EX  \\\n",
       "0        1    1014727   G   T    1013497    1013576  1230         0   \n",
       "1        1    1021242   C   T    1020120    1020373  1122         0   \n",
       "2        1    1208352   G   A    1206571    1206385 -1781         0   \n",
       "3        1    1360908   G   A    1358543    1358456 -2365         0   \n",
       "4        1    1398713   C   T    1399311    1399019   598         0   \n",
       "...    ...        ...  ..  ..        ...        ...   ...       ...   \n",
       "2945     X  151976628   G   A  151974676  151974570 -1952         0   \n",
       "2946     X  152735883   C   T  152733735  152733668 -2148         0   \n",
       "2947     X  152735883   C   T  152733757  152733859  2126         0   \n",
       "2948     X  153727239   G   A  153724856  153726166  2383         0   \n",
       "2949     X  154749911   A   G  154751077  154750954  1166         0   \n",
       "\n",
       "      DOWNSTREAM  INTERGENIC STRAND           ENSGID           ENSTID  \\\n",
       "0              1           0      +  ENSG00000187608  ENST00000649529   \n",
       "1              1           0      +  ENSG00000188157  ENST00000379370   \n",
       "2              0           1      -  ENSG00000186891  ENST00000328596   \n",
       "3              0           1      -  ENSG00000162576  ENST00000445648   \n",
       "4              1           0      -  ENSG00000221978  ENST00000481223   \n",
       "...          ...         ...    ...              ...              ...   \n",
       "2945           0           1      -  ENSG00000102287  ENST00000370328   \n",
       "2946           0           1      -  ENSG00000198930  ENST00000370287   \n",
       "2947           1           0      +  ENSG00000213401  ENST00000393869   \n",
       "2948           1           0      +  ENSG00000101986  ENST00000218104   \n",
       "2949           1           0      -  ENSG00000160219  ENST00000424127   \n",
       "\n",
       "                                                 WINDOW  \n",
       "0     CCTCGGTCTCTCAAAGTGCTGGGATTACAGGTGAGCCACTGCGCCC...  \n",
       "1     ATCTGGCCCCACCCACATCCTGCTGATGGGTAGAGCCTAGTGGTCT...  \n",
       "2     GGGGGCGTCGTATCCCTTCAGTCTCCGGCAGGACAGGGGGTAGGTC...  \n",
       "3     ACAGAAGTACACGTACGTACACATACACAGACACAGAAGCGCACAC...  \n",
       "4     GGTGGGACCGAGGGGAGTTCCCGACGCACGTGTCACGGGCGAAAGG...  \n",
       "...                                                 ...  \n",
       "2945  GAGACAGTTTGTGGTGTGTATCGACCAAAGTTGTGGGACCGACCGA...  \n",
       "2946  GTTCGGGTCCCGGAGAGGAACCCGGAGTTCCGGAAGGAGTCCGAAC...  \n",
       "2947  TAATAAAACGTTTTTTATCACAAATATGCCATTAAAAGTGAAATCC...  \n",
       "2948  TTCATTTGCTTTTCCAGTTTCCTTTGTGTAGAAATCAGATTGTCTG...  \n",
       "2949  AGGAAACGGTTTACAATAAAAAAGGGTAAGTCGTCTCTTGTGATCT...  \n",
       "\n",
       "[2950 rows x 14 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_matches_columns = [\t# column headers\n",
    "\t'CHROM',\n",
    "\t'POS',\n",
    "\t'REF',\n",
    "\t'ALT',\n",
    "\t'TSS',\n",
    "\t'FIRST_EIB',\n",
    "\t'DIST',\n",
    "\t'FIRST_EX',\n",
    "\t'DOWNSTREAM',\n",
    "\t'INTERGENIC',\n",
    "\t'STRAND',\n",
    "\t'ENSGID',\n",
    "\t'ENSTID',\n",
    "\t'WINDOW',\n",
    "]\n",
    "\n",
    "tss_matches = pd.DataFrame(\n",
    "\ttss_matches_list,\n",
    "\tcolumns=tss_matches_columns,\n",
    ")\n",
    "\n",
    "# ensure dtypes for CHROM, TSS, FIRST_EIB, and DIST\n",
    "tss_matches.CHROM = tss_matches.CHROM.astype(str)\n",
    "tss_matches.TSS = tss_matches.TSS.astype(int)\n",
    "tss_matches.FIRST_EIB = tss_matches.FIRST_EIB.astype(int)\n",
    "tss_matches.DIST = tss_matches.DIST.astype(int)\n",
    "\n",
    "tss_matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{MATCHES_DIR}/tss_matches.csv\", WRITING_MODE) as fout:\n",
    "\ttss_matches.to_csv(\n",
    "\t\tfout,\n",
    "\t\tindex=False,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **3.2** Intergenic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-section finds matches for the intergenic coordinates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_intergen_matches(windows: pd.DataFrame, dataset: pd.DataFrame, win_len: int=None, print_progress: bool=True) -> list:\n",
    "\t'''\n",
    "\tGet DNMs (\"matches\") that fall within the windows specified by `windows`.\n",
    "\n",
    "\t## Parameters\n",
    "\twindows: Pandas.DataFrame\n",
    "\t\tWindows around specific coordinates.\n",
    "\tdataset: Pandas.DataFrame\n",
    "\t\tDNMs dataset.\n",
    "\twin_len: int, default = None\n",
    "\t\tWindow length to be used to consider DNMs. Can be less than or equal \n",
    "\t\tto length of the window sequence associated with each coordinate in `windows`. \n",
    "\t\tIf none is specified, defaults to length of window sequence from `windows`.\n",
    "\tprint_progress: bool, default = True\n",
    "\t\tIf true, will print progress as percentage of completion.\n",
    "\t\t\n",
    "\t## Returns\n",
    "\tmatches: list\n",
    "\t\tList with each item containing the following information for each coordinate:\n",
    "\n",
    "\t\t\t- chromosome\n",
    "\t\t\t- DNM position\n",
    "\t\t\t- reference base\n",
    "\t\t\t- alternate base\n",
    "\t\t\t- coordinate for center of window\n",
    "\t\t\t- distance from center\n",
    "\t\t\t- window sequence\n",
    "\t'''\n",
    "\n",
    "\tif win_len == None:\n",
    "\t\twin_len = len(windows.WINDOW.iloc[0])\n",
    "\n",
    "\tmatches = list()\n",
    "\tfor i, coord in enumerate(windows.values):\n",
    "\t\tchrom, center, window = coord\n",
    "\n",
    "\t\ttemp_matches = dataset[\n",
    "\t\t\t(dataset.CHROM == chrom) &\n",
    "\t\t\t(dataset.POS >= (center - win_len//2)) &\n",
    "\t\t\t(dataset.POS <= (center + win_len//2))\n",
    "\t\t]\n",
    "\n",
    "\t\tif len(temp_matches):\n",
    "\t\t\tmatches.extend([ (chrom, pos, ref, alt, center, (pos - center), window)\n",
    "\t\t\t\t\t\t\t\t\tfor chrom, pos, ref, alt\n",
    "\t\t\t\t\t\t\t\t\tin temp_matches.values ])\n",
    "\n",
    "\t\tif print_progress: print(f\"{i * 100 / len(windows):.0f}%\", end='\\r')\n",
    "\n",
    "\treturn matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergen_match_columns = [\n",
    "\t'CHROM',\n",
    "\t'POS',\n",
    "\t'REF',\n",
    "\t'ALT',\n",
    "\t'CENTER',\n",
    "\t'DIST',\n",
    "\t'WINDOW',\n",
    "]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **3.2.1** GC matched"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-sub-section finds matches for the GC matched intergenic coordinates. Matches are aggregatedc in a list (`intergen_GCmatch_match_list`) and subsequently loaded into a Pandas DataFrame (`intergen_GCmatch_matches`).\n",
    "\n",
    "Each match is packed with the following metrics:\n",
    "\n",
    "|Metric|Description|\n",
    "|-:|-|\n",
    "|`CHROM`|chromosome of the DNM|\n",
    "|`POS`|DNM position on chromosome CHROM|\n",
    "|`REF`|base at position POS in reference genome|\n",
    "|`ALT`|base found at position POS in proband|\n",
    "|`CENTER`|coordinate for center of window|\n",
    "|`DIST`|distance from CENTER|\n",
    "|`WINDOW`|string representing ±2501 base window centered around CENTER|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*NOTE: `INTERGEN_GCMATCH_WINLEN` is used to generate these matches, which is different from `WINDOW_LENGTH`. The window sequence represented by `WINDOW` was generated using `WINDOW_LENGTH`, and thus extends beyond `INTERGEN_GCMATCH_WINLEN` bases from `CENTER`*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4%\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100%\r"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>CENTER</th>\n",
       "      <th>DIST</th>\n",
       "      <th>WINDOW</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69216275</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>69216254</td>\n",
       "      <td>21</td>\n",
       "      <td>CTATTACCATGAGAATAATGTGTCTGGAGCAACCTGTGACAGTTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>171081744</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>171081712</td>\n",
       "      <td>32</td>\n",
       "      <td>AACCCAAATGTCCAACAATGATAGACTGGATTAAGAAAATGTGGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>238429890</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>238429898</td>\n",
       "      <td>-8</td>\n",
       "      <td>TTTTGGCCGTTCTTGCAGGAGTGAGATGGTATCACATTGTGGTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>5297395</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>5297374</td>\n",
       "      <td>21</td>\n",
       "      <td>AAAGCATGGTCTTAGTTGTATCTTTTCAATTCCCAGGCAAAAGAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>54914884</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>54914879</td>\n",
       "      <td>5</td>\n",
       "      <td>ATCTCTACCAAAAATACAAAAAATTAGCCGGGCATGGTGGTGGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>6</td>\n",
       "      <td>17206664</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>17206686</td>\n",
       "      <td>-22</td>\n",
       "      <td>TGCTTCCATTTGATGTTTTGTTTTGTTTTTCTTGTAACTGAGTCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>6</td>\n",
       "      <td>44063154</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>44063164</td>\n",
       "      <td>-10</td>\n",
       "      <td>GGATTCCTATAAAGTAAGGATACCCGTGGACCCCACCTCACCTGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>1</td>\n",
       "      <td>218165004</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>218165029</td>\n",
       "      <td>-25</td>\n",
       "      <td>AATATGGGTATAAACCAAAAATGTTTTGGTTTTCTTCTGTTTAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>7</td>\n",
       "      <td>63114344</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>63114327</td>\n",
       "      <td>17</td>\n",
       "      <td>TTTGTATTTTTTAGTAGACACAGGGTTTCACCGTGTTAACCAGGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>13</td>\n",
       "      <td>109311826</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>109311801</td>\n",
       "      <td>25</td>\n",
       "      <td>ATCTGGTCATATTTATACTCAGTTAAGTACTTGACTACTAAATTGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS REF ALT     CENTER  DIST  \\\n",
       "0      1   69216275   A   T   69216254    21   \n",
       "1      1  171081744   T   A  171081712    32   \n",
       "2      1  238429890   C   T  238429898    -8   \n",
       "3      1    5297395   C   A    5297374    21   \n",
       "4      1   54914884   A   C   54914879     5   \n",
       "..   ...        ...  ..  ..        ...   ...   \n",
       "67     6   17206664   G   A   17206686   -22   \n",
       "68     6   44063154   G   A   44063164   -10   \n",
       "69     1  218165004   C   G  218165029   -25   \n",
       "70     7   63114344   G   C   63114327    17   \n",
       "71    13  109311826   C   T  109311801    25   \n",
       "\n",
       "                                               WINDOW  \n",
       "0   CTATTACCATGAGAATAATGTGTCTGGAGCAACCTGTGACAGTTTG...  \n",
       "1   AACCCAAATGTCCAACAATGATAGACTGGATTAAGAAAATGTGGCA...  \n",
       "2   TTTTGGCCGTTCTTGCAGGAGTGAGATGGTATCACATTGTGGTTTT...  \n",
       "3   AAAGCATGGTCTTAGTTGTATCTTTTCAATTCCCAGGCAAAAGAGT...  \n",
       "4   ATCTCTACCAAAAATACAAAAAATTAGCCGGGCATGGTGGTGGGAG...  \n",
       "..                                                ...  \n",
       "67  TGCTTCCATTTGATGTTTTGTTTTGTTTTTCTTGTAACTGAGTCTC...  \n",
       "68  GGATTCCTATAAAGTAAGGATACCCGTGGACCCCACCTCACCTGGC...  \n",
       "69  AATATGGGTATAAACCAAAAATGTTTTGGTTTTCTTCTGTTTAAAG...  \n",
       "70  TTTGTATTTTTTAGTAGACACAGGGTTTCACCGTGTTAACCAGGAA...  \n",
       "71  ATCTGGTCATATTTATACTCAGTTAAGTACTTGACTACTAAATTGT...  \n",
       "\n",
       "[72 rows x 7 columns]"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergen_GCmatch_match = get_intergen_matches(\n",
    "\twindows=intergen_GCmatch_windows,\n",
    "    dataset=iceland_dataset,\n",
    "\twin_len=INTERGEN_GCMATCH_WINLEN,\n",
    ")\n",
    "intergen_GCmatch_matches = pd.DataFrame(\n",
    "\tintergen_GCmatch_match,\n",
    "\tcolumns=intergen_match_columns,\n",
    ")\n",
    "intergen_GCmatch_matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{MATCHES_DIR}/intergen_GCmatch_matches_100b.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_GCmatch_matches.to_csv(\n",
    "\t\tfout,\n",
    "\t\tindex=False,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **3.2.2** non-GC matched"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-sub-section finds matches for the non-GC matched intergenic coordinates, employing the same method as 3.2.1 (see chart in 3.2.1 for summary of metrics). Matches are aggregatedc in a list (`intergen_nonGCmatch_match_list`) and subsequently loaded into a Pandas DataFrame (`intergen_nonGCmatch_matches`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0%\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100%\r"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>CENTER</th>\n",
       "      <th>DIST</th>\n",
       "      <th>WINDOW</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69216275</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>69215104</td>\n",
       "      <td>1171</td>\n",
       "      <td>CAAAAGAAGGGGGGAGGTCTCTTCCTTTGTTTGCAATAGGGAGTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>171476920</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>171475608</td>\n",
       "      <td>1312</td>\n",
       "      <td>CACAGTTCCACATGACTGGAGAGGCCTCACAATCATGGCAGAAGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>171081744</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>171081362</td>\n",
       "      <td>382</td>\n",
       "      <td>AAAATGCTCACCATCACTGGCCATCAGAGAAATGCAAATCAAAACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>192455849</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>192455397</td>\n",
       "      <td>452</td>\n",
       "      <td>TAACTTCACTACCTGATAGCTCTGTGATCTTGGATATGTTTCTAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>238429890</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>238427448</td>\n",
       "      <td>2442</td>\n",
       "      <td>ACTTACATGTGAAAAAAATCTTGAAATTATATTTTGAATTTTCATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28869</th>\n",
       "      <td>22</td>\n",
       "      <td>23675643</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>23673219</td>\n",
       "      <td>2424</td>\n",
       "      <td>GGGCAAGGAGCAGGGTCAGACCCAGCCTCAGACACTCCCTGGCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28870</th>\n",
       "      <td>22</td>\n",
       "      <td>34300244</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>34300075</td>\n",
       "      <td>169</td>\n",
       "      <td>CAGAGCCCAGAAGAAGAACTGCTGAGCTCTGAGTCCTGGACTTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28871</th>\n",
       "      <td>22</td>\n",
       "      <td>47662694</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>47660885</td>\n",
       "      <td>1809</td>\n",
       "      <td>CCCTCCACCAGGAATGCCTTTGCCCCTTTGTGCATGGCCAGGTCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28872</th>\n",
       "      <td>22</td>\n",
       "      <td>29305002</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>29303607</td>\n",
       "      <td>1395</td>\n",
       "      <td>CTCTCAGTCACCAACCACCTCTAATGTTGCCCAATCAATGGTCTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28873</th>\n",
       "      <td>22</td>\n",
       "      <td>30632678</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>30632617</td>\n",
       "      <td>61</td>\n",
       "      <td>CCCCTCTGGGCTCCCATGGAGGATGGTGCTGCTGGGTCACCTGCTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28874 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM        POS REF ALT     CENTER  DIST  \\\n",
       "0         1   69216275   A   T   69215104  1171   \n",
       "1         1  171476920   T   A  171475608  1312   \n",
       "2         1  171081744   T   A  171081362   382   \n",
       "3         1  192455849   A   C  192455397   452   \n",
       "4         1  238429890   C   T  238427448  2442   \n",
       "...     ...        ...  ..  ..        ...   ...   \n",
       "28869    22   23675643   G   A   23673219  2424   \n",
       "28870    22   34300244   A   G   34300075   169   \n",
       "28871    22   47662694   G   A   47660885  1809   \n",
       "28872    22   29305002   G   A   29303607  1395   \n",
       "28873    22   30632678   A   G   30632617    61   \n",
       "\n",
       "                                                  WINDOW  \n",
       "0      CAAAAGAAGGGGGGAGGTCTCTTCCTTTGTTTGCAATAGGGAGTAT...  \n",
       "1      CACAGTTCCACATGACTGGAGAGGCCTCACAATCATGGCAGAAGGC...  \n",
       "2      AAAATGCTCACCATCACTGGCCATCAGAGAAATGCAAATCAAAACC...  \n",
       "3      TAACTTCACTACCTGATAGCTCTGTGATCTTGGATATGTTTCTAAA...  \n",
       "4      ACTTACATGTGAAAAAAATCTTGAAATTATATTTTGAATTTTCATT...  \n",
       "...                                                  ...  \n",
       "28869  GGGCAAGGAGCAGGGTCAGACCCAGCCTCAGACACTCCCTGGCACC...  \n",
       "28870  CAGAGCCCAGAAGAAGAACTGCTGAGCTCTGAGTCCTGGACTTTTT...  \n",
       "28871  CCCTCCACCAGGAATGCCTTTGCCCCTTTGTGCATGGCCAGGTCCA...  \n",
       "28872  CTCTCAGTCACCAACCACCTCTAATGTTGCCCAATCAATGGTCTTT...  \n",
       "28873  CCCCTCTGGGCTCCCATGGAGGATGGTGCTGCTGGGTCACCTGCTT...  \n",
       "\n",
       "[28874 rows x 7 columns]"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergen_nonGCmatch_matches = get_intergen_matches(\n",
    "\twindows=intergen_nonGCmatch_windows,\n",
    "    dataset=iceland_dataset,\n",
    "\twin_len=WINDOW_LENGTH,\n",
    ")\n",
    "intergen_nonGCmatch_matches = pd.DataFrame(\n",
    "\tintergen_nonGCmatch_matches,\n",
    "\tcolumns=intergen_match_columns,\n",
    ")\n",
    "intergen_nonGCmatch_matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{MATCHES_DIR}/intergen_nonGCmatch_matches.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_nonGCmatch_matches.to_csv(\n",
    "\t\tfout,\n",
    "\t\tindex=False,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "______"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **4.** <u>Separate CpG and non-CpG</u>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This section separates CpG and non-CpG DNMs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [],
   "source": [
    "def separate_CpG_matches(matches: pd.DataFrame, ref_idx: int, alt_idx: int, dist_idx: int, win_len: int, window_idx: int=-1, offset: int=0) -> tuple:\n",
    "    '''\n",
    "    Separate CpG and non-CpG DNMs.\n",
    "\n",
    "    ## Parameters\n",
    "    matches: Pandas.DataFrame\n",
    "        DNM DataFrame.\n",
    "    ref_idx: int\n",
    "        REF column index.\n",
    "    alt_idx: int\n",
    "        ALT column index.\n",
    "    dist_idx: int\n",
    "        DIST column index.\n",
    "    win_len: int\n",
    "        Window length to be used to consider DNMs. Can be less than or equal \n",
    "        to length of the window sequence associated with each match in `matches`.\n",
    "    window_idx: int, default = -1\n",
    "        WINDOW column index.\n",
    "    offset: int, default = 0\n",
    "        Optional offset for positional indexing within the window.\n",
    "        \n",
    "    ## Returns\n",
    "    separated: tuple\n",
    "        Tuple containing two Pandas.DataFrames, the first being the CpG DNMs and \n",
    "        the second the non-CpG DNMs.\n",
    "    '''\n",
    "\n",
    "    CpG_matches_list, nonCpG_matches_list = list(), list()\n",
    "    for dnm in matches.values:\n",
    "        ref = dnm[ref_idx]\n",
    "        alt = dnm[alt_idx]\n",
    "        dist = dnm[dist_idx]\n",
    "        window = dnm[window_idx]\n",
    "\n",
    "        if   ref == 'C' and alt == 'T' and window[win_len//2+int(dist)+1+offset] == 'G':   # CpG C>T\n",
    "            CpG_matches_list.append(dnm)\n",
    "        elif ref == 'G' and alt == 'A' and window[win_len//2+int(dist)-1+offset] == 'C':   # CpG G>A\n",
    "            CpG_matches_list.append(dnm)\n",
    "        else: nonCpG_matches_list.append(dnm)\n",
    "\n",
    "    CpG_matches = pd.DataFrame(CpG_matches_list, columns=matches.columns)\n",
    "    nonCpG_matches = pd.DataFrame(nonCpG_matches_list, columns=matches.columns)\n",
    "\n",
    "    return CpG_matches, nonCpG_matches"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **4.1** TSS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-section separates CpG and non-CpG DNMs around the TSS into 2 Pandas DataFrames:\n",
    "- `tss_CpG`: CpG DNMs around the TSS\n",
    "- `tss_nonCpG`: non-CpG DNMs around the TSS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(tss_CpG) = 426, len(tss_nonCpG) = 2524\n"
     ]
    }
   ],
   "source": [
    "tss_CpG, tss_nonCpG = separate_CpG_matches(\n",
    "    matches=tss_matches,\n",
    "    ref_idx=2,\n",
    "    alt_idx=3,\n",
    "    dist_idx=6,\n",
    "    win_len=WINDOW_LENGTH,\n",
    "    offset=CPG_OFFSET,\n",
    ")\n",
    "print(f\"{len(tss_CpG) = }, {len(tss_nonCpG) = }\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{MATCHES_DIR}/tss_CpG.csv\", WRITING_MODE) as fout:\n",
    "\ttss_CpG.to_csv(\n",
    "\t\tfout,\n",
    "\t\tindex=False,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)\n",
    "\n",
    "with open(f\"{MATCHES_DIR}/tss_nonCpG.csv\", WRITING_MODE) as fout:\n",
    "\ttss_nonCpG.to_csv(\n",
    "\t\tfout,\n",
    "\t\tindex=False,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **4.2** Intergenic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-section separactes CpG and non-CpG DNMs around the intergenic coordinates."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **4.2.1** GC matched"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-sub-section separactes CpG and non-CpG DNMs around the intergenic GC matched coordinates into 2 Pandas DataFrames:\n",
    "- `intergen_GCmatch_CpG`: CpG DNMs around the intergenic GC matched coordinates\n",
    "- `intergen_GCmatch_nonCpG`: non-CpG DNMs around the intergenic GC matched coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(intergen_GCmatch_CpG) = 32, len(intergen_GCmatch_nonCpG) = 40\n"
     ]
    }
   ],
   "source": [
    "intergen_GCmatch_CpG, intergen_GCmatch_nonCpG = separate_CpG_matches(\n",
    "    matches=intergen_GCmatch_matches,\n",
    "    ref_idx=2,\n",
    "    alt_idx=3,\n",
    "    dist_idx=5,\n",
    "    win_len=WINDOW_LENGTH,\n",
    ")\n",
    "print(f\"{len(intergen_GCmatch_CpG) = }, {len(intergen_GCmatch_nonCpG) = }\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{MATCHES_DIR}/intergen_GCmatch_CpG_100b.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_GCmatch_CpG.to_csv(\n",
    "\t\tfout,\n",
    "\t\tindex=False,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)\n",
    "\n",
    "with open(f\"{MATCHES_DIR}/intergen_GCmatch_nonCpG_100b.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_GCmatch_nonCpG.to_csv(\n",
    "\t\tfout,\n",
    "\t\tindex=False,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **4.2.2** non-GC matched"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-sub-section separactes CpG and non-CpG DNMs around the intergenic non-GC matched coordinates into 2 Pandas DataFrames:\n",
    "- `intergen_nonGCmatch_CpG`: CpG DNMs around the intergenic non-GC matched coordinates\n",
    "- `intergen_nonGCmatch_nonCpG`: non-CpG DNMs around the intergenic non-GC matched coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len(intergen_nonGCmatch_CpG) = 4748, len(intergen_nonGCmatch_nonCpG) = 24126\n"
     ]
    }
   ],
   "source": [
    "intergen_nonGCmatch_CpG, intergen_nonGCmatch_nonCpG = separate_CpG_matches(\n",
    "    matches=intergen_nonGCmatch_matches,\n",
    "    ref_idx=2,\n",
    "    alt_idx=3,\n",
    "    dist_idx=5,\n",
    "    win_len=WINDOW_LENGTH,\n",
    ")\n",
    "print(f\"{len(intergen_nonGCmatch_CpG) = }, {len(intergen_nonGCmatch_nonCpG) = }\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{MATCHES_DIR}/intergen_nonGCmatch_CpG.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_nonGCmatch_CpG.to_csv(\n",
    "\t\tfout,\n",
    "\t\tindex=False,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)\n",
    "\n",
    "with open(f\"{MATCHES_DIR}/intergen_nonGCmatch_nonCpG.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_nonGCmatch_nonCpG.to_csv(\n",
    "\t\tfout,\n",
    "\t\tindex=False,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "________"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **5.** <u>Positional counts</u>"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This section generates positional counts for each base. That is, this counts the number of occurances of each base in each unique position in the ±2501 base window.\n",
    "\n",
    "Each subsection generates 3 sets of positional counts:\n",
    "- total positional counts (CpG + non-CpG)\n",
    "- positional counts excluding CpGs\n",
    "- C and G positional counts within CpGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_total_pos_counts(windows: pd.DataFrame, win_len: int, offset: int=0, column_headers: list[str]=None) -> pd.DataFrame:\n",
    "    '''\n",
    "    Get positional counts for each base within windows specified by `windows`.\n",
    "\n",
    "    ## Parameters\n",
    "    windows: Pandas.DataFrame\n",
    "        Windows around coordinates.\n",
    "    win_len: int\n",
    "        Length of window.\n",
    "    offset: int, default = 0\n",
    "        Window offset.\n",
    "    column_headers: list, default = None\n",
    "        If none is specified, defaults to ['A','C','G','T'].\n",
    "        \n",
    "    ## Returns\n",
    "    matches: Pandas.DataFrame\n",
    "        DataFrame containing positional counts for each base. \n",
    "        Rows are position relative to center of window, and \n",
    "        columns are the bases.\n",
    "    '''\n",
    "\n",
    "    if column_headers == None:\n",
    "        column_headers = ['A','C','G','T']\n",
    "\n",
    "    # grab windows from coordinate DataFrame\n",
    "    windows_listed = np.array([ list(window[offset:-offset])\n",
    "                                    for window\n",
    "                                    in windows.WINDOW ]).transpose()\n",
    "\n",
    "    # count unique occurences of each base\n",
    "    pos_counts = np.array([ np.unique(row, return_counts=True)[1]\n",
    "                                for row\n",
    "                                in windows_listed ])\n",
    "\n",
    "    return pd.DataFrame(\n",
    "        pos_counts,\n",
    "        columns=column_headers,\n",
    "        index=range(-win_len//2,win_len//2),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_CpG_pos_counts(windows: pd.DataFrame, win_len: int, CpG_str: str='CG', offset: int=1, column_headers: list[str]=None) -> pd.DataFrame:\n",
    "\t'''\n",
    "\tGet positional counts for each CpG within windows specified by `windows`.\n",
    "\n",
    "\t## Parameters\n",
    "\twindows: Pandas.DataFrame\n",
    "\t\tWindows around coordinates.\n",
    "\twin_len: int\n",
    "\t\tLength of window.\n",
    "\tCpG_str: str, default = 'CG'\n",
    "\t\tString representation of a CpG.\n",
    "\toffset: int, default = 1\n",
    "\t\tOffsets C counts and G counts, relative to each other.\n",
    "\tcolumn_headers: list, default = None\n",
    "\t\tIf none is specified, defaults to ['A','C','G','T'].\n",
    "\t\t\n",
    "\t## Returns\n",
    "\tCpG_pos_counts: Pandas.DataFrame\n",
    "\t\tDataFrame containing positional counts for each base. \n",
    "\t\tRows are position relative to center of window, and \n",
    "\t\tcolumns are the bases.\n",
    "\t'''\n",
    "\n",
    "\tif column_headers == None:\n",
    "\t\tcolumn_headers = ['A','C','G','T']\n",
    "\n",
    "\t# grab CpGs (used for non-CpG and CpG positional counts)\n",
    "\tCpG_pos_counts_raw = [ [CpG_match.start() for CpG_match in re.finditer(CpG_str, i)]\n",
    "\t\t\t\t\t\t\t\tfor i\n",
    "\t\t\t\t\t\t\t\tin windows.WINDOW.iloc ]\n",
    "\n",
    "\t# flatten raw counts\n",
    "\tCpG_pos_counts_flattened = np.array([ num\n",
    "\t\t\t\t\t\t\t\t\t\t\tfor item\n",
    "\t\t\t\t\t\t\t\t\t\t\tin CpG_pos_counts_raw\n",
    "\t\t\t\t\t\t\t\t\t\t\tfor num in (item if isinstance(item, list) else (item,)) ])\n",
    "\n",
    "\t# offset raw counts\n",
    "\tCpG_pos_counts_flattened -= (win_len//2 + 1)\n",
    "\n",
    "\n",
    "\tCpG_pos_counts_dict = OrderedDict(sorted(\n",
    "\t\tCounter(CpG_pos_counts_flattened).items()\n",
    "\t))\n",
    "\n",
    "\treturn pd.DataFrame(\n",
    "\t\t[\n",
    "\t\t\tnp.zeros(win_len),\t\t\t\t\t\t\t\t\t\t# 'A' count (zero)\n",
    "\t\t\tnp.array([*CpG_pos_counts_dict.values(),][offset:]),\t# 'C' count\n",
    "\t\t\tnp.array([*CpG_pos_counts_dict.values(),][:-offset]),\t# 'G' count\n",
    "\t\t\tnp.zeros(win_len),\t\t\t\t\t\t\t\t\t\t# 'T' count (zero)\n",
    "\t\t],\n",
    "\t\tcolumns=range(-win_len//2,win_len//2),\n",
    "\t\tindex=column_headers,\n",
    "\t\tdtype=int,\n",
    "\t).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nonCpG_pos_counts(total_pos_counts: pd.DataFrame, CpG_pos_counts: pd.DataFrame, offset: int=1, column_headers: list[str]=None) -> pd.DataFrame:\n",
    "\t'''\n",
    "\tGet positional counts for each non-CpG base within windows specified by `windows`.\n",
    "\n",
    "\t## Parameters\n",
    "\ttotal_pos_counts: Pandas.DataFrame\n",
    "\t\tTotal positional counts.\n",
    "\tCpG_pos_counts: Pandas.DataFrame\n",
    "\t\tCpG positional counts.\n",
    "\toffset: int, default = 1\n",
    "\t\tOffsets C counts and G counts, relative to each other.\n",
    "\tcolumn_headers: list, default = None\n",
    "\t\tIf none is specified, defaults to ['A','C','G','T'].\n",
    "\t\t\n",
    "\t## Returns\n",
    "\tnonCpG_pos_counts: Pandas.DataFrame\n",
    "\t\tDataFrame containing positional counts for each base. \n",
    "\t\tRows are position relative to center of window, and \n",
    "\t\tcolumns are the bases.\n",
    "\t'''\n",
    "\n",
    "\tif column_headers == None:\n",
    "\t\tcolumn_headers = ['A','C','G','T']\n",
    "\n",
    "\tnonCpG_C_pos_counts = total_pos_counts.C - np.array([np.zeros(offset),*CpG_pos_counts.C][offset:])\n",
    "\tnonCpG_G_pos_counts = total_pos_counts.G - np.array([*CpG_pos_counts.G,np.zeros(offset)][:-offset])\n",
    "\n",
    "\treturn pd.DataFrame(\n",
    "\t\t[\n",
    "\t\t\ttotal_pos_counts['A'],\t# 'A' count\n",
    "\t\t\tnonCpG_C_pos_counts,\t# 'C' count\n",
    "\t\t\tnonCpG_G_pos_counts,\t# 'G' count\n",
    "\t\t\ttotal_pos_counts['T'],\t# 'T' count\n",
    "\t\t],\n",
    "\t\tcolumns=total_pos_counts.index,\n",
    "\t\tindex=total_pos_counts.columns,\n",
    "\t\tdtype=int,\n",
    "\t).T"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **5.1** TSS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-section generates positional counts for the TSS coordinates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-2500</th>\n",
       "      <td>4653</td>\n",
       "      <td>3802</td>\n",
       "      <td>3865</td>\n",
       "      <td>4784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2499</th>\n",
       "      <td>4717</td>\n",
       "      <td>3855</td>\n",
       "      <td>3913</td>\n",
       "      <td>4619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2498</th>\n",
       "      <td>4674</td>\n",
       "      <td>3818</td>\n",
       "      <td>3894</td>\n",
       "      <td>4718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2497</th>\n",
       "      <td>4753</td>\n",
       "      <td>3840</td>\n",
       "      <td>3870</td>\n",
       "      <td>4641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2496</th>\n",
       "      <td>4746</td>\n",
       "      <td>3872</td>\n",
       "      <td>3880</td>\n",
       "      <td>4606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2495</th>\n",
       "      <td>4648</td>\n",
       "      <td>3827</td>\n",
       "      <td>3919</td>\n",
       "      <td>4710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2496</th>\n",
       "      <td>4660</td>\n",
       "      <td>3874</td>\n",
       "      <td>3894</td>\n",
       "      <td>4676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2497</th>\n",
       "      <td>4594</td>\n",
       "      <td>3931</td>\n",
       "      <td>3864</td>\n",
       "      <td>4715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>4720</td>\n",
       "      <td>3828</td>\n",
       "      <td>3793</td>\n",
       "      <td>4763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2499</th>\n",
       "      <td>4646</td>\n",
       "      <td>3876</td>\n",
       "      <td>3893</td>\n",
       "      <td>4689</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          A     C     G     T\n",
       "-2500  4653  3802  3865  4784\n",
       "-2499  4717  3855  3913  4619\n",
       "-2498  4674  3818  3894  4718\n",
       "-2497  4753  3840  3870  4641\n",
       "-2496  4746  3872  3880  4606\n",
       "...     ...   ...   ...   ...\n",
       " 2495  4648  3827  3919  4710\n",
       " 2496  4660  3874  3894  4676\n",
       " 2497  4594  3931  3864  4715\n",
       " 2498  4720  3828  3793  4763\n",
       " 2499  4646  3876  3893  4689\n",
       "\n",
       "[5000 rows x 4 columns]"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# total positional counts\n",
    "\n",
    "tss_pos_counts = get_total_pos_counts(\n",
    "    windows=tss_windows,\n",
    "    win_len=WINDOW_LENGTH,\n",
    "    offset=CPG_OFFSET,\n",
    ")\n",
    "tss_pos_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-2500</th>\n",
       "      <td>0</td>\n",
       "      <td>553</td>\n",
       "      <td>537</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2499</th>\n",
       "      <td>0</td>\n",
       "      <td>534</td>\n",
       "      <td>553</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2498</th>\n",
       "      <td>0</td>\n",
       "      <td>563</td>\n",
       "      <td>534</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2497</th>\n",
       "      <td>0</td>\n",
       "      <td>594</td>\n",
       "      <td>563</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2496</th>\n",
       "      <td>0</td>\n",
       "      <td>551</td>\n",
       "      <td>594</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2495</th>\n",
       "      <td>0</td>\n",
       "      <td>562</td>\n",
       "      <td>535</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2496</th>\n",
       "      <td>0</td>\n",
       "      <td>546</td>\n",
       "      <td>562</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2497</th>\n",
       "      <td>0</td>\n",
       "      <td>519</td>\n",
       "      <td>546</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>0</td>\n",
       "      <td>549</td>\n",
       "      <td>519</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2499</th>\n",
       "      <td>0</td>\n",
       "      <td>597</td>\n",
       "      <td>549</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       A    C    G  T\n",
       "-2500  0  553  537  0\n",
       "-2499  0  534  553  0\n",
       "-2498  0  563  534  0\n",
       "-2497  0  594  563  0\n",
       "-2496  0  551  594  0\n",
       "...   ..  ...  ... ..\n",
       " 2495  0  562  535  0\n",
       " 2496  0  546  562  0\n",
       " 2497  0  519  546  0\n",
       " 2498  0  549  519  0\n",
       " 2499  0  597  549  0\n",
       "\n",
       "[5000 rows x 4 columns]"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CpG positional counts\n",
    "\n",
    "tss_CpG_pos_counts = get_CpG_pos_counts(\n",
    "\twindows=tss_windows,\n",
    "\twin_len=WINDOW_LENGTH,\n",
    ")\n",
    "tss_CpG_pos_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-2500</th>\n",
       "      <td>4653</td>\n",
       "      <td>3249</td>\n",
       "      <td>3328</td>\n",
       "      <td>4784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2499</th>\n",
       "      <td>4717</td>\n",
       "      <td>3321</td>\n",
       "      <td>3360</td>\n",
       "      <td>4619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2498</th>\n",
       "      <td>4674</td>\n",
       "      <td>3255</td>\n",
       "      <td>3360</td>\n",
       "      <td>4718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2497</th>\n",
       "      <td>4753</td>\n",
       "      <td>3246</td>\n",
       "      <td>3307</td>\n",
       "      <td>4641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2496</th>\n",
       "      <td>4746</td>\n",
       "      <td>3321</td>\n",
       "      <td>3286</td>\n",
       "      <td>4606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2495</th>\n",
       "      <td>4648</td>\n",
       "      <td>3265</td>\n",
       "      <td>3384</td>\n",
       "      <td>4710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2496</th>\n",
       "      <td>4660</td>\n",
       "      <td>3328</td>\n",
       "      <td>3332</td>\n",
       "      <td>4676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2497</th>\n",
       "      <td>4594</td>\n",
       "      <td>3412</td>\n",
       "      <td>3318</td>\n",
       "      <td>4715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>4720</td>\n",
       "      <td>3279</td>\n",
       "      <td>3274</td>\n",
       "      <td>4763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2499</th>\n",
       "      <td>4646</td>\n",
       "      <td>3279</td>\n",
       "      <td>3344</td>\n",
       "      <td>4689</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          A     C     G     T\n",
       "-2500  4653  3249  3328  4784\n",
       "-2499  4717  3321  3360  4619\n",
       "-2498  4674  3255  3360  4718\n",
       "-2497  4753  3246  3307  4641\n",
       "-2496  4746  3321  3286  4606\n",
       "...     ...   ...   ...   ...\n",
       " 2495  4648  3265  3384  4710\n",
       " 2496  4660  3328  3332  4676\n",
       " 2497  4594  3412  3318  4715\n",
       " 2498  4720  3279  3274  4763\n",
       " 2499  4646  3279  3344  4689\n",
       "\n",
       "[5000 rows x 4 columns]"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# non-CpG positional counts\n",
    "\n",
    "tss_nonCpG_pos_counts = get_nonCpG_pos_counts(\n",
    "\ttotal_pos_counts=tss_pos_counts,\n",
    "\tCpG_pos_counts=tss_CpG_pos_counts,\n",
    ")\n",
    "tss_nonCpG_pos_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{POS_COUNTS_SUBDIR}/tss_pos_counts.csv\", WRITING_MODE) as fout:\n",
    "\ttss_pos_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)\n",
    "with open(f\"{POS_COUNTS_SUBDIR}/tss_nonCpG_pos_counts.csv\", WRITING_MODE) as fout:\n",
    "\ttss_nonCpG_pos_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)\n",
    "with open(f\"{POS_COUNTS_SUBDIR}/tss_CpG_pos_counts.csv\", WRITING_MODE) as fout:\n",
    "\ttss_CpG_pos_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **5.2** Intergenic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-section generates positional counts for the intergenic coordinates."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **5.2.1** GC matched"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-sub-section generates positional counts for the intergenic GC matched coordinates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-2500</th>\n",
       "      <td>4601</td>\n",
       "      <td>3761</td>\n",
       "      <td>3769</td>\n",
       "      <td>4437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2499</th>\n",
       "      <td>4612</td>\n",
       "      <td>3732</td>\n",
       "      <td>3797</td>\n",
       "      <td>4427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2498</th>\n",
       "      <td>4537</td>\n",
       "      <td>3809</td>\n",
       "      <td>3746</td>\n",
       "      <td>4476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2497</th>\n",
       "      <td>4519</td>\n",
       "      <td>3743</td>\n",
       "      <td>3771</td>\n",
       "      <td>4535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2496</th>\n",
       "      <td>4418</td>\n",
       "      <td>3738</td>\n",
       "      <td>3948</td>\n",
       "      <td>4464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2495</th>\n",
       "      <td>4523</td>\n",
       "      <td>3842</td>\n",
       "      <td>3727</td>\n",
       "      <td>4476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2496</th>\n",
       "      <td>4412</td>\n",
       "      <td>3770</td>\n",
       "      <td>3735</td>\n",
       "      <td>4651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2497</th>\n",
       "      <td>4343</td>\n",
       "      <td>3813</td>\n",
       "      <td>3795</td>\n",
       "      <td>4617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>4533</td>\n",
       "      <td>3861</td>\n",
       "      <td>3783</td>\n",
       "      <td>4391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2499</th>\n",
       "      <td>4506</td>\n",
       "      <td>3818</td>\n",
       "      <td>3755</td>\n",
       "      <td>4489</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          A     C     G     T\n",
       "-2500  4601  3761  3769  4437\n",
       "-2499  4612  3732  3797  4427\n",
       "-2498  4537  3809  3746  4476\n",
       "-2497  4519  3743  3771  4535\n",
       "-2496  4418  3738  3948  4464\n",
       "...     ...   ...   ...   ...\n",
       " 2495  4523  3842  3727  4476\n",
       " 2496  4412  3770  3735  4651\n",
       " 2497  4343  3813  3795  4617\n",
       " 2498  4533  3861  3783  4391\n",
       " 2499  4506  3818  3755  4489\n",
       "\n",
       "[5000 rows x 4 columns]"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# total positional counts\n",
    "\n",
    "intergen_GCmatch_pos_counts = get_total_pos_counts(\n",
    "    windows=intergen_GCmatch_windows,\n",
    "    win_len=WINDOW_LENGTH,\n",
    "    offset=CPG_OFFSET,\n",
    ")\n",
    "intergen_GCmatch_pos_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-2500</th>\n",
       "      <td>0</td>\n",
       "      <td>250</td>\n",
       "      <td>236</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2499</th>\n",
       "      <td>0</td>\n",
       "      <td>240</td>\n",
       "      <td>250</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2498</th>\n",
       "      <td>0</td>\n",
       "      <td>237</td>\n",
       "      <td>240</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2497</th>\n",
       "      <td>0</td>\n",
       "      <td>259</td>\n",
       "      <td>237</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2496</th>\n",
       "      <td>0</td>\n",
       "      <td>247</td>\n",
       "      <td>259</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2495</th>\n",
       "      <td>0</td>\n",
       "      <td>247</td>\n",
       "      <td>230</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2496</th>\n",
       "      <td>0</td>\n",
       "      <td>239</td>\n",
       "      <td>247</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2497</th>\n",
       "      <td>0</td>\n",
       "      <td>235</td>\n",
       "      <td>239</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>0</td>\n",
       "      <td>246</td>\n",
       "      <td>235</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2499</th>\n",
       "      <td>0</td>\n",
       "      <td>254</td>\n",
       "      <td>246</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       A    C    G  T\n",
       "-2500  0  250  236  0\n",
       "-2499  0  240  250  0\n",
       "-2498  0  237  240  0\n",
       "-2497  0  259  237  0\n",
       "-2496  0  247  259  0\n",
       "...   ..  ...  ... ..\n",
       " 2495  0  247  230  0\n",
       " 2496  0  239  247  0\n",
       " 2497  0  235  239  0\n",
       " 2498  0  246  235  0\n",
       " 2499  0  254  246  0\n",
       "\n",
       "[5000 rows x 4 columns]"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CpG positional counts\n",
    "\n",
    "intergen_GCmatch_CpG_pos_counts = get_CpG_pos_counts(\n",
    "\twindows=intergen_GCmatch_windows,\n",
    "\twin_len=WINDOW_LENGTH,\n",
    ")\n",
    "intergen_GCmatch_CpG_pos_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-2500</th>\n",
       "      <td>4601</td>\n",
       "      <td>3511</td>\n",
       "      <td>3533</td>\n",
       "      <td>4437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2499</th>\n",
       "      <td>4612</td>\n",
       "      <td>3492</td>\n",
       "      <td>3547</td>\n",
       "      <td>4427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2498</th>\n",
       "      <td>4537</td>\n",
       "      <td>3572</td>\n",
       "      <td>3506</td>\n",
       "      <td>4476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2497</th>\n",
       "      <td>4519</td>\n",
       "      <td>3484</td>\n",
       "      <td>3534</td>\n",
       "      <td>4535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2496</th>\n",
       "      <td>4418</td>\n",
       "      <td>3491</td>\n",
       "      <td>3689</td>\n",
       "      <td>4464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2495</th>\n",
       "      <td>4523</td>\n",
       "      <td>3595</td>\n",
       "      <td>3497</td>\n",
       "      <td>4476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2496</th>\n",
       "      <td>4412</td>\n",
       "      <td>3531</td>\n",
       "      <td>3488</td>\n",
       "      <td>4651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2497</th>\n",
       "      <td>4343</td>\n",
       "      <td>3578</td>\n",
       "      <td>3556</td>\n",
       "      <td>4617</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>4533</td>\n",
       "      <td>3615</td>\n",
       "      <td>3548</td>\n",
       "      <td>4391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2499</th>\n",
       "      <td>4506</td>\n",
       "      <td>3564</td>\n",
       "      <td>3509</td>\n",
       "      <td>4489</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          A     C     G     T\n",
       "-2500  4601  3511  3533  4437\n",
       "-2499  4612  3492  3547  4427\n",
       "-2498  4537  3572  3506  4476\n",
       "-2497  4519  3484  3534  4535\n",
       "-2496  4418  3491  3689  4464\n",
       "...     ...   ...   ...   ...\n",
       " 2495  4523  3595  3497  4476\n",
       " 2496  4412  3531  3488  4651\n",
       " 2497  4343  3578  3556  4617\n",
       " 2498  4533  3615  3548  4391\n",
       " 2499  4506  3564  3509  4489\n",
       "\n",
       "[5000 rows x 4 columns]"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# non-CpG positional counts\n",
    "\n",
    "intergen_GCmatch_nonCpG_pos_counts = get_nonCpG_pos_counts(\n",
    "\ttotal_pos_counts=intergen_GCmatch_pos_counts,\n",
    "\tCpG_pos_counts=intergen_GCmatch_CpG_pos_counts,\n",
    ")\n",
    "intergen_GCmatch_nonCpG_pos_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{POS_COUNTS_SUBDIR}/intergen_GCmatch_pos_counts.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_GCmatch_pos_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)\n",
    "with open(f\"{POS_COUNTS_SUBDIR}/intergen_GCmatch_nonCpG_pos_counts.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_GCmatch_nonCpG_pos_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)\n",
    "with open(f\"{POS_COUNTS_SUBDIR}/intergen_GCmatch_CpG_pos_counts.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_GCmatch_CpG_pos_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **5.2.2** non-GC matched"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This sub-sub-section generates positional counts for the intergenic GC matched coordinates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-2500</th>\n",
       "      <td>49224</td>\n",
       "      <td>32580</td>\n",
       "      <td>33004</td>\n",
       "      <td>48700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2499</th>\n",
       "      <td>48765</td>\n",
       "      <td>32642</td>\n",
       "      <td>33171</td>\n",
       "      <td>48930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2498</th>\n",
       "      <td>48774</td>\n",
       "      <td>32637</td>\n",
       "      <td>32874</td>\n",
       "      <td>49223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2497</th>\n",
       "      <td>49072</td>\n",
       "      <td>32798</td>\n",
       "      <td>32623</td>\n",
       "      <td>49015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2496</th>\n",
       "      <td>48686</td>\n",
       "      <td>32994</td>\n",
       "      <td>33049</td>\n",
       "      <td>48779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2495</th>\n",
       "      <td>48612</td>\n",
       "      <td>32497</td>\n",
       "      <td>33069</td>\n",
       "      <td>49330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2496</th>\n",
       "      <td>49170</td>\n",
       "      <td>33018</td>\n",
       "      <td>32741</td>\n",
       "      <td>48579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2497</th>\n",
       "      <td>48749</td>\n",
       "      <td>32576</td>\n",
       "      <td>32812</td>\n",
       "      <td>49371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>48809</td>\n",
       "      <td>32529</td>\n",
       "      <td>32996</td>\n",
       "      <td>49174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2499</th>\n",
       "      <td>48881</td>\n",
       "      <td>32615</td>\n",
       "      <td>32902</td>\n",
       "      <td>49110</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           A      C      G      T\n",
       "-2500  49224  32580  33004  48700\n",
       "-2499  48765  32642  33171  48930\n",
       "-2498  48774  32637  32874  49223\n",
       "-2497  49072  32798  32623  49015\n",
       "-2496  48686  32994  33049  48779\n",
       "...      ...    ...    ...    ...\n",
       " 2495  48612  32497  33069  49330\n",
       " 2496  49170  33018  32741  48579\n",
       " 2497  48749  32576  32812  49371\n",
       " 2498  48809  32529  32996  49174\n",
       " 2499  48881  32615  32902  49110\n",
       "\n",
       "[5000 rows x 4 columns]"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# total positional counts\n",
    "\n",
    "intergen_nonGCmatch_pos_counts = get_total_pos_counts(\n",
    "    windows=intergen_nonGCmatch_windows,\n",
    "    win_len=WINDOW_LENGTH,\n",
    "    offset=CPG_OFFSET,\n",
    ")\n",
    "intergen_nonGCmatch_pos_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-2500</th>\n",
       "      <td>0</td>\n",
       "      <td>1544</td>\n",
       "      <td>1493</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2499</th>\n",
       "      <td>0</td>\n",
       "      <td>1434</td>\n",
       "      <td>1544</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2498</th>\n",
       "      <td>0</td>\n",
       "      <td>1472</td>\n",
       "      <td>1434</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2497</th>\n",
       "      <td>0</td>\n",
       "      <td>1517</td>\n",
       "      <td>1472</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2496</th>\n",
       "      <td>0</td>\n",
       "      <td>1553</td>\n",
       "      <td>1517</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2495</th>\n",
       "      <td>0</td>\n",
       "      <td>1469</td>\n",
       "      <td>1552</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2496</th>\n",
       "      <td>0</td>\n",
       "      <td>1454</td>\n",
       "      <td>1469</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2497</th>\n",
       "      <td>0</td>\n",
       "      <td>1440</td>\n",
       "      <td>1454</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>0</td>\n",
       "      <td>1469</td>\n",
       "      <td>1440</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2499</th>\n",
       "      <td>0</td>\n",
       "      <td>1495</td>\n",
       "      <td>1469</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       A     C     G  T\n",
       "-2500  0  1544  1493  0\n",
       "-2499  0  1434  1544  0\n",
       "-2498  0  1472  1434  0\n",
       "-2497  0  1517  1472  0\n",
       "-2496  0  1553  1517  0\n",
       "...   ..   ...   ... ..\n",
       " 2495  0  1469  1552  0\n",
       " 2496  0  1454  1469  0\n",
       " 2497  0  1440  1454  0\n",
       " 2498  0  1469  1440  0\n",
       " 2499  0  1495  1469  0\n",
       "\n",
       "[5000 rows x 4 columns]"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CpG positional counts\n",
    "\n",
    "intergen_nonGCmatch_CpG_pos_counts = get_CpG_pos_counts(\n",
    "\twindows=intergen_nonGCmatch_windows,\n",
    "\twin_len=WINDOW_LENGTH,\n",
    ")\n",
    "intergen_nonGCmatch_CpG_pos_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-2500</th>\n",
       "      <td>49224</td>\n",
       "      <td>31036</td>\n",
       "      <td>31511</td>\n",
       "      <td>48700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2499</th>\n",
       "      <td>48765</td>\n",
       "      <td>31208</td>\n",
       "      <td>31627</td>\n",
       "      <td>48930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2498</th>\n",
       "      <td>48774</td>\n",
       "      <td>31165</td>\n",
       "      <td>31440</td>\n",
       "      <td>49223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2497</th>\n",
       "      <td>49072</td>\n",
       "      <td>31281</td>\n",
       "      <td>31151</td>\n",
       "      <td>49015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-2496</th>\n",
       "      <td>48686</td>\n",
       "      <td>31441</td>\n",
       "      <td>31532</td>\n",
       "      <td>48779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2495</th>\n",
       "      <td>48612</td>\n",
       "      <td>31028</td>\n",
       "      <td>31517</td>\n",
       "      <td>49330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2496</th>\n",
       "      <td>49170</td>\n",
       "      <td>31564</td>\n",
       "      <td>31272</td>\n",
       "      <td>48579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2497</th>\n",
       "      <td>48749</td>\n",
       "      <td>31136</td>\n",
       "      <td>31358</td>\n",
       "      <td>49371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>48809</td>\n",
       "      <td>31060</td>\n",
       "      <td>31556</td>\n",
       "      <td>49174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2499</th>\n",
       "      <td>48881</td>\n",
       "      <td>31120</td>\n",
       "      <td>31433</td>\n",
       "      <td>49110</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           A      C      G      T\n",
       "-2500  49224  31036  31511  48700\n",
       "-2499  48765  31208  31627  48930\n",
       "-2498  48774  31165  31440  49223\n",
       "-2497  49072  31281  31151  49015\n",
       "-2496  48686  31441  31532  48779\n",
       "...      ...    ...    ...    ...\n",
       " 2495  48612  31028  31517  49330\n",
       " 2496  49170  31564  31272  48579\n",
       " 2497  48749  31136  31358  49371\n",
       " 2498  48809  31060  31556  49174\n",
       " 2499  48881  31120  31433  49110\n",
       "\n",
       "[5000 rows x 4 columns]"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# non-CpG positional counts\n",
    "\n",
    "intergen_nonGCmatch_nonCpG_pos_counts = get_nonCpG_pos_counts(\n",
    "\ttotal_pos_counts=intergen_nonGCmatch_pos_counts,\n",
    "\tCpG_pos_counts=intergen_nonGCmatch_CpG_pos_counts,\n",
    ")\n",
    "intergen_nonGCmatch_nonCpG_pos_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{POS_COUNTS_SUBDIR}/intergen_nonGCmatch_pos_counts.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_nonGCmatch_pos_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)\n",
    "with open(f\"{POS_COUNTS_SUBDIR}/intergen_nonGCmatch_nonCpG_pos_counts.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_nonGCmatch_nonCpG_pos_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)\n",
    "with open(f\"{POS_COUNTS_SUBDIR}/intergen_nonGCmatch_CpG_pos_counts.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_nonGCmatch_CpG_pos_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "____"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **6.** <u>Regional counts (TSS only)</u>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_regional_counts_regions = [\n",
    "\t'INTERGEN',\n",
    "\t'FIRST_EX',\n",
    "\t'DOWNSTREAM',\n",
    "]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **6.1** Count instances of mutable bases in each region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>T</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>INTERGEN</th>\n",
       "      <td>10884639</td>\n",
       "      <td>10894242</td>\n",
       "      <td>10495883</td>\n",
       "      <td>10485236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FIRST_EX</th>\n",
       "      <td>915616</td>\n",
       "      <td>910596</td>\n",
       "      <td>1620611</td>\n",
       "      <td>1622955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DOWNSTREAM</th>\n",
       "      <td>9530273</td>\n",
       "      <td>9570171</td>\n",
       "      <td>9294443</td>\n",
       "      <td>9295335</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   A         T         C         G\n",
       "INTERGEN    10884639  10894242  10495883  10485236\n",
       "FIRST_EX      915616    910596   1620611   1622955\n",
       "DOWNSTREAM   9530273   9570171   9294443   9295335"
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# generate dict of bases and counts for each region\n",
    "tss_counts = [ dict(zip(BASES_SET, np.zeros(len(BASES_SET), dtype=int)))\n",
    "\t\t\t  \tfor i\n",
    "\t\t\t\tin range(len(tss_regional_counts_regions)) ]\n",
    "\n",
    "# instantiate global CpG counters for each region\n",
    "intergen_CpG_count, first_ex_CpG_count, downstream_CpG_count = 0, 0, 0\n",
    "\n",
    "\n",
    "for match in tss_windows.values:\n",
    "\t_, tss, eib, _, _, _, window = match\n",
    "\n",
    "\t# calculate positional index for exon-intron boundary\n",
    "\teib_index = WINDOW_LENGTH//2 + abs(eib - tss)\n",
    "\t\n",
    "\t# get sequence for each region\n",
    "\tintergenic_seq = window[ : WINDOW_LENGTH//2 + 1 + CPG_OFFSET]\n",
    "\tfirst_ex_seq = window[WINDOW_LENGTH//2 - 1 + CPG_OFFSET : eib_index + 1 + CPG_OFFSET]\n",
    "\tdownstream_seq = window[eib_index - 1 + CPG_OFFSET : ]\n",
    "\n",
    "\t# count CpGs in each region\n",
    "\tintergen_CpG_count += len(re.findall(CPG_STR,intergenic_seq))\n",
    "\tfirst_ex_CpG_count += len(re.findall(CPG_STR,first_ex_seq))\n",
    "\tdownstream_CpG_count += len(re.findall(CPG_STR,downstream_seq))\n",
    "\n",
    "\t# count number of mutable bases in each region\n",
    "\tcounts_temp = [ Counter(seq[CPG_OFFSET:-CPG_OFFSET])\n",
    "\t\t\t\t\t\tfor seq\n",
    "\t\t\t\t\t\tin (intergenic_seq, first_ex_seq, downstream_seq) ]\n",
    "\n",
    "\t# store mutable bases counts in global counter\n",
    "\tfor i, counts in enumerate(counts_temp):\n",
    "\t\tfor base, base_count in counts.items():\n",
    "\t\t\ttss_counts[i][base] += base_count\n",
    "\n",
    "\n",
    "tss_region_counts = pd.DataFrame(\n",
    "\ttss_counts,\n",
    "\tindex=tss_regional_counts_regions,\n",
    ")\n",
    "tss_region_counts = tss_region_counts[['A','T','C','G']]\t# explicitly reorder\n",
    "\n",
    "tss_region_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>T</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>INTERGEN</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1896839</td>\n",
       "      <td>1896839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FIRST_EX</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>492596</td>\n",
       "      <td>492596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DOWNSTREAM</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1616800</td>\n",
       "      <td>1616800</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            A  T        C        G\n",
       "INTERGEN    0  0  1896839  1896839\n",
       "FIRST_EX    0  0   492596   492596\n",
       "DOWNSTREAM  0  0  1616800  1616800"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CpG_region_counts = intergen_CpG_count, first_ex_CpG_count, downstream_CpG_count\n",
    "\n",
    "# simple aggregation of regional CpG counts\n",
    "tss_CpG_region_counts = pd.DataFrame(\n",
    "\t[\n",
    "        np.zeros(len(tss_regional_counts_regions), dtype=int),\t# A\n",
    "        np.zeros(len(tss_regional_counts_regions), dtype=int),\t# T\n",
    "        CpG_region_counts,\t\t\t\t\t\t\t\t\t\t# C\n",
    "        CpG_region_counts,\t\t\t\t\t\t\t\t\t\t# G\n",
    "    ],\n",
    "\tcolumns=tss_region_counts.index,\n",
    "\tindex=tss_region_counts.columns,\n",
    ").T\n",
    "\n",
    "tss_CpG_region_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A</th>\n",
       "      <th>T</th>\n",
       "      <th>C</th>\n",
       "      <th>G</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>INTERGEN</th>\n",
       "      <td>10884639</td>\n",
       "      <td>10894242</td>\n",
       "      <td>8599044</td>\n",
       "      <td>8588397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FIRST_EX</th>\n",
       "      <td>915616</td>\n",
       "      <td>910596</td>\n",
       "      <td>1128015</td>\n",
       "      <td>1130359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DOWNSTREAM</th>\n",
       "      <td>9530273</td>\n",
       "      <td>9570171</td>\n",
       "      <td>7677643</td>\n",
       "      <td>7678535</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   A         T        C        G\n",
       "INTERGEN    10884639  10894242  8599044  8588397\n",
       "FIRST_EX      915616    910596  1128015  1130359\n",
       "DOWNSTREAM   9530273   9570171  7677643  7678535"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subtract regional CpG counts from regional mutable C and G regional counts\n",
    "nonCpG_CG_region_counts = [\t[tss_counts[i][base] - CpG_region_counts[i] for i in range(len(tss_regional_counts_regions))]\n",
    "\t\t\t\t\t\t\t\tfor base\n",
    "\t\t\t\t\t\t\t\tin CPG_STR ]\n",
    "\n",
    "nonCpG_region_counts = [\n",
    "\ttss_region_counts['A'].to_list(),\t# A\n",
    "\ttss_region_counts['T'].to_list(),\t# T\n",
    "\tnonCpG_CG_region_counts[0],\t\t\t# C\n",
    "\tnonCpG_CG_region_counts[1],\t\t\t# G\n",
    "]\n",
    "\n",
    "tss_nonCpG_region_counts = pd.DataFrame(\n",
    "\tnonCpG_region_counts,\n",
    "\tcolumns=tss_regional_counts_regions,\n",
    "\tindex=tss_region_counts.columns,\n",
    ").T\n",
    "tss_nonCpG_region_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{COUNTS_DIR}/regional/tss_region_counts.csv\", WRITING_MODE) as fout:\n",
    "\ttss_region_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)\n",
    "\n",
    "with open(f\"{COUNTS_DIR}/regional/tss_nonCpG_region_counts.csv\", WRITING_MODE) as fout:\n",
    "\ttss_nonCpG_region_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)\n",
    "\n",
    "with open(f\"{COUNTS_DIR}/regional/tss_CpG_region_counts.csv\", WRITING_MODE) as fout:\n",
    "\ttss_CpG_region_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **6.2** Count instances of unique mutation type in each region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_regional_dnm_type_counts(matches: pd.DataFrame, dnm_types: set[str], regions: list[str]) -> pd.DataFrame:\n",
    "\t'''\n",
    "\tGet regional DNM type counts for each DNM type (i.e. 'C>T', 'A>C') in `matches`.\n",
    "\n",
    "\t## Parameters\n",
    "\tmatches: Pandas.DataFrame\n",
    "\t\tDataFrame containing DNM dataset.\n",
    "\tdnm_types: set\n",
    "\t\tSet of all possible DNM types.\n",
    "\tregions: list\n",
    "\t\tList of strings representing region names.\n",
    "\t## Returns\n",
    "\tcounts: Pandas.DataFrame\n",
    "\t\tDataFrame containing regional counts for each DNM type. \n",
    "\t\tRows are regions, and columns are the DNM types.\n",
    "\t'''\n",
    "\n",
    "\tcounts = [ get_ordered_counts_for_region(matches=matches, dnm_types=dnm_types, region=region)\n",
    "\t\t\t\tfor region\n",
    "\t\t\t\tin regions ]\n",
    "\n",
    "\treturn pd.DataFrame(\n",
    "\t\tcounts,\n",
    "\t\tindex=regions,\n",
    "\t)\n",
    "\n",
    "def get_ordered_counts_for_region(matches: pd.DataFrame, region: str, dnm_types: set[str]) -> OrderedDict:\n",
    "\t'''\n",
    "\tGet DNM type counts for each DNM type in the region specified by `region`.\n",
    "\n",
    "\t## Parameters\n",
    "\tmatches: Pandas.DataFrame\n",
    "\t\tDataFrame containing DNM dataset.\n",
    "\tregion: str\n",
    "\t\tStrings representing region name.\n",
    "\tdnm_types: set\n",
    "\t\tSet of all possible DNM types.\n",
    "\t## Returns\n",
    "\tordered_counts: OrderedDict\n",
    "\t\tOrderedDict containing all possible DNM types \n",
    "\t\tand their respective counts for the region.\n",
    "\t'''\n",
    "\n",
    "\tcounts = Counter([ '>'.join(ref_alt)\n",
    "\t\t\t\t\t\tfor ref_alt\n",
    "\t\t\t\t\t\tin matches[matches[region].astype(int) == 1][['REF','ALT']].values ])\n",
    "\tset_zero_values(counts=counts, dnm_types=dnm_types)\n",
    "\n",
    "\treturn OrderedDict(sorted(counts.items()))\n",
    "\n",
    "def set_zero_values(counts: Counter, dnm_types: set[str]) -> None:\n",
    "\t'''\n",
    "\tAdds missing DNM types to `counts`, and sets their \n",
    "\trespective count to zero.\n",
    "\n",
    "\t## Parameters\n",
    "\tcounts: Counter\n",
    "\t\tCounter containing DNM types and their respective counts.\n",
    "\tdnm_types: set\n",
    "\t\tSet of all possible DNM types.\n",
    "\t'''\n",
    "\n",
    "\tcounts.update({ dnm_type : 0\n",
    "\t\t\t\t\t\tfor dnm_type\n",
    "\t\t\t\t\t\tin dnm_types - set(counts.keys()) })\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **6.2.1** Total (non-CpG + CpG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A&gt;C</th>\n",
       "      <th>A&gt;G</th>\n",
       "      <th>A&gt;T</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>G&gt;A</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>T&gt;G</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>INTERGENIC</th>\n",
       "      <td>41</td>\n",
       "      <td>141</td>\n",
       "      <td>36</td>\n",
       "      <td>48</td>\n",
       "      <td>96</td>\n",
       "      <td>331</td>\n",
       "      <td>340</td>\n",
       "      <td>87</td>\n",
       "      <td>52</td>\n",
       "      <td>39</td>\n",
       "      <td>149</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FIRST_EX</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DOWNSTREAM</th>\n",
       "      <td>34</td>\n",
       "      <td>146</td>\n",
       "      <td>36</td>\n",
       "      <td>66</td>\n",
       "      <td>80</td>\n",
       "      <td>301</td>\n",
       "      <td>300</td>\n",
       "      <td>107</td>\n",
       "      <td>56</td>\n",
       "      <td>30</td>\n",
       "      <td>183</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            A>C  A>G  A>T  C>A  C>G  C>T  G>A  G>C  G>T  T>A  T>C  T>G\n",
       "INTERGENIC   41  141   36   48   96  331  340   87   52   39  149   36\n",
       "FIRST_EX      0    2    1    1    4    6    6    0    0    0    4    0\n",
       "DOWNSTREAM   34  146   36   66   80  301  300  107   56   30  183   44"
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_dnmtype_counts = get_regional_dnm_type_counts(\n",
    "\tmatches=tss_matches,\n",
    "\tdnm_types=DNM_TYPES,\n",
    "\tregions=tss_regional_counts_regions,\n",
    ")\n",
    "tss_dnmtype_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{COUNTS_DIR}/tss_dnmtype_counts.csv\", WRITING_MODE) as fout:\n",
    "\ttss_dnmtype_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **6.2.2** non-CpG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A&gt;C</th>\n",
       "      <th>A&gt;G</th>\n",
       "      <th>A&gt;T</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>G&gt;A</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>T&gt;G</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>INTERGENIC</th>\n",
       "      <td>41</td>\n",
       "      <td>141</td>\n",
       "      <td>36</td>\n",
       "      <td>48</td>\n",
       "      <td>96</td>\n",
       "      <td>224</td>\n",
       "      <td>238</td>\n",
       "      <td>87</td>\n",
       "      <td>52</td>\n",
       "      <td>39</td>\n",
       "      <td>149</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FIRST_EX</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DOWNSTREAM</th>\n",
       "      <td>34</td>\n",
       "      <td>146</td>\n",
       "      <td>36</td>\n",
       "      <td>66</td>\n",
       "      <td>80</td>\n",
       "      <td>192</td>\n",
       "      <td>221</td>\n",
       "      <td>107</td>\n",
       "      <td>56</td>\n",
       "      <td>30</td>\n",
       "      <td>183</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            A>C  A>G  A>T  C>A  C>G  C>T  G>A  G>C  G>T  T>A  T>C  T>G\n",
       "INTERGENIC   41  141   36   48   96  224  238   87   52   39  149   36\n",
       "FIRST_EX      0    2    1    1    4    1    6    0    0    0    4    0\n",
       "DOWNSTREAM   34  146   36   66   80  192  221  107   56   30  183   44"
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_nonCpG_dnmtype_counts = get_regional_dnm_type_counts(\n",
    "\tmatches=tss_nonCpG,\n",
    "\tdnm_types=DNM_TYPES,\n",
    "\tregions=tss_regional_counts_regions,\n",
    ")\n",
    "tss_nonCpG_dnmtype_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{COUNTS_DIR}/tss_nonCpG_dnmtype_counts.csv\", WRITING_MODE) as fout:\n",
    "\ttss_nonCpG_dnmtype_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **6.2.2** CpG"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Note: this is a redundant calculation as these values can be obtained by simply subtracting the non-CpG counts from the total counts. This redundant calculation is done to verify the previous results.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A&gt;C</th>\n",
       "      <th>A&gt;G</th>\n",
       "      <th>A&gt;T</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>G&gt;A</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>T&gt;G</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>INTERGENIC</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>107</td>\n",
       "      <td>102</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FIRST_EX</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DOWNSTREAM</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>109</td>\n",
       "      <td>79</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            A>C  A>G  A>T  C>A  C>G  C>T  G>A  G>C  G>T  T>A  T>C  T>G\n",
       "INTERGENIC    0    0    0    0    0  107  102    0    0    0    0    0\n",
       "FIRST_EX      0    0    0    0    0    5    0    0    0    0    0    0\n",
       "DOWNSTREAM    0    0    0    0    0  109   79    0    0    0    0    0"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_CpG_dnmtype_counts = get_regional_dnm_type_counts(\n",
    "\tmatches=tss_CpG,\n",
    "\tdnm_types=DNM_TYPES,\n",
    "\tregions=tss_regional_counts_regions,\n",
    ")\n",
    "tss_CpG_dnmtype_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out results\n",
    "\n",
    "with open(f\"{COUNTS_DIR}/tss_CpG_dnmtype_counts.csv\", WRITING_MODE) as fout:\n",
    "\ttss_CpG_dnmtype_counts.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "___________"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **7.** <u>Positional Metrics</u>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_positional_metrics(df: pd.DataFrame, total_pos_counts: pd.DataFrame, nonCpG_pos_counts:pd.DataFrame, win_len: int, sliding_win_len: int, GC_loss_dnm_types: list, include_CpG: bool, column_headers: list, parent_dir: str, writing_mode: str='x') -> None:\n",
    "    '''\n",
    "    Print to a CSV file the positional DNM type metrics for DNM types involving the loss or gain of GC content (i.e. 'C>T', 'A>G').\n",
    "\n",
    "    Each set of metrics includes:\n",
    "    |Metric|Description|\n",
    "    |-|-|\n",
    "    |Number of DNMs|Number of occurences of the DNM type in a given position|\n",
    "    |Average of DNMs|Sliding window average for number of the DNM|\n",
    "    |Positional count|Positional count of mutable bases (i.e. REF)|\n",
    "    |Rate of mutation|Positional rate of the DNM|\n",
    "    |Average rate of mutation|Sliding window average for rate of mutation|\n",
    "\n",
    "    Filenames will have the structure \"ref\" + \"alt\", where \"ref\" is the reference base \n",
    "    in a given DNM and \"alt\" is the base found in a proband. For example, for A>G DNMs, \n",
    "    the filename will be \"AG.csv\".\n",
    "\n",
    "    ## Parameters\n",
    "    df: Pandas.DataFrame\n",
    "        Pandas DataFrame containing DNMs dataset.\n",
    "    total_pos_counts: Pandas.DataFrame\n",
    "        Pandas DataFrame containing total positional base counts (CpG + nonCpG).\n",
    "    nonCpG_pos_counts: Pandas.DataFrame\n",
    "        Pandas DataFrame containing nonCpG positional base counts.\n",
    "    win_len: int\n",
    "        Length of sequence window.\n",
    "    sliding_win_len: int\n",
    "        Length of sliding window used to generate Average of DNMs and Average rate of mutation.\n",
    "    GC_loss_dnm_types: set\n",
    "        Set of possible GC DNM types.\n",
    "    include_CpG: bool\n",
    "        Whether to include CpG DNMs in counts.\n",
    "    column_headers: list\n",
    "        Headers used for printed columns.\n",
    "    parent_dir: str\n",
    "        Parent directory to write output files.\n",
    "    writing_mode: str, default = 'x'\n",
    "        Whether to overwrite existing files. Default is 'x', which does not overwrite existing files. \n",
    "        If set to 'w', will overwrite existing files.\n",
    "    '''\n",
    "\n",
    "    for ref, alt in GC_loss_dnm_types:\n",
    "        dnm_counts_array = get_dnm_counts(\n",
    "            df=df,\n",
    "            ref=ref,\n",
    "            alt=alt,\n",
    "            win_len=win_len,\n",
    "        )\n",
    "\n",
    "        selected_pos_counts = select_pos_counts(\n",
    "            ref=ref,\n",
    "            alt=alt,\n",
    "            include_CpG=include_CpG,\n",
    "            total_pos_counts=total_pos_counts,\n",
    "            nonCpG_pos_counts=nonCpG_pos_counts,\n",
    "        )\n",
    "\n",
    "        avg_dnm_counts_array = sliding_window_avg(\n",
    "            arr=dnm_counts_array,\n",
    "            sliding_win_len=sliding_win_len,\n",
    "        )\n",
    "\n",
    "        mut_rate_array = dnm_counts_array / selected_pos_counts\n",
    "\n",
    "        avg_mut_rate_array = sliding_window_avg(\n",
    "            arr=mut_rate_array,\n",
    "            sliding_win_len=sliding_win_len,\n",
    "        )\n",
    "\n",
    "        df_tmp = pd.DataFrame(\n",
    "            data = [\n",
    "                dnm_counts_array,\n",
    "                avg_dnm_counts_array,\n",
    "                selected_pos_counts,\n",
    "                mut_rate_array,\n",
    "                avg_mut_rate_array,\n",
    "            ],\n",
    "        ).T.convert_dtypes()\n",
    "        df_tmp.columns = column_headers\n",
    "        df_tmp.index = range(-win_len//2, win_len//2)\n",
    "\n",
    "        if not os.path.exists(parent_dir):\n",
    "            os.makedirs(parent_dir)\n",
    "\n",
    "        # write out\n",
    "        with open(f\"{parent_dir}/{ref}{alt}.csv\", writing_mode) as fout:\n",
    "            df_tmp.to_csv(\n",
    "                fout,\n",
    "                float_format='%.10f',\n",
    "                lineterminator='\\n',\n",
    "            )\n",
    "\n",
    "\n",
    "def print_positional_netGC(df: pd.DataFrame, pos_counts:pd.DataFrame, win_len: int, GC_loss_dnm_types: list, column_headers: list, parent_dir: str, sliding_win_len: int=100, truncated_output: bool=False, writing_mode: str='x') -> None:\n",
    "\t'''\n",
    "\tPrint to a CSV file the net GC metrics for DNM types involving the loss or gain of GC content (i.e. 'C>T', 'A>G').\n",
    "\n",
    "\tIf `truncated_output` is `False`, each set of metrics includes:\n",
    "\t|Metric|Description|\n",
    "\t|-|-|\n",
    "\t|Number of DNMs|Net number of occurences of the DNM type in a given position|\n",
    "\t|Average of DNMs|Sliding window average for net number of the DNM|\n",
    "\t|Positional count|Positional count of C and G|\n",
    "\t|Rate of mutation|Net positional rate of the DNM|\n",
    "\t|Average rate of mutation|Sliding window average for net positional rate of mutation|\n",
    "\n",
    "\tFilename will be \"netGC.csv\".\n",
    "\n",
    "\tIf `truncated_output` is `False`, each set of metrics includes:\n",
    "\t|Metric|Description|\n",
    "\t|-|-|\n",
    "\t|Net GC change|Net positional change in GC|\n",
    "\t|Positional count|Positional count of C and G|\n",
    "\n",
    "\tFilename will be \"netGC_#b.csv\", where \"#\" is the window size specified in `win_len`.\n",
    "\n",
    "\t## Parameters\n",
    "\tdf: Pandas.DataFrame\n",
    "\t\tPandas DataFrame containing DNMs dataset.\n",
    "\tpos_counts: Pandas.DataFrame\n",
    "\t\tPandas DataFrame containing positional base counts.\n",
    "\twin_len: int\n",
    "\t\tLength of sequence window.\n",
    "\tGC_loss_dnm_types: set\n",
    "\t\tSet of possible GC DNM types.\n",
    "\tcolumn_headers: list\n",
    "\t\tHeaders used for printed columns.\n",
    "\tparent_dir: str\n",
    "\t\tParent directory to write output files.\n",
    "\tsliding_win_len: int, default = 100\n",
    "\t\tLength of sliding window used to generate Average of DNMs and Average rate of mutation.\n",
    "\ttruncated_output: bool, default = False\n",
    "\t\tWhether output will be truncated.\n",
    "\twriting_mode: str, default = 'x'\n",
    "\t\tWhether to overwrite existing files. Default is 'x', which does not overwrite existing files. \n",
    "\t\tIf set to 'w', will overwrite existing files.\n",
    "\t'''\n",
    "\n",
    "    \n",
    "\t# get positional DNM counts for all GC gain/loss DNM types\n",
    "\traw_counts = [ get_dnm_counts(df=df, ref=ref, alt=alt, win_len=win_len)\n",
    "\t\t\t\t\tfor ref, alt\n",
    "\t\t\t\t\tin GC_loss_dnm_types ]\n",
    "\n",
    "\t# calculate net GC change\n",
    "\t#                                   net GC gain                                                       net GC loss\n",
    "\tnet_CG = (raw_counts[0] + raw_counts[1] + raw_counts[2] + raw_counts[3]) - (raw_counts[4] + raw_counts[5] + raw_counts[6] + raw_counts[7])\n",
    "\n",
    "\tif not truncated_output:\n",
    "\t\tpos_counts_C = pos_counts.C.to_numpy()\n",
    "\t\tpos_counts_G = pos_counts.G.to_numpy()\n",
    "\n",
    "\t\t# get average GC change using sliding window averaging\n",
    "\t\tavg_CG = sliding_window_avg(\n",
    "\t\t\tarr=net_CG,\n",
    "\t\t\tsliding_win_len=sliding_win_len,\n",
    "\t\t)\n",
    "\n",
    "\t\t# calculate net GC change rate\n",
    "\t\tCG_rate = net_CG / (pos_counts_C + pos_counts_G)\n",
    "\n",
    "\t\t# calculate average GC change rate using sliding window averaging\n",
    "\t\tavg_CG_rate = sliding_window_avg(\n",
    "\t\t\tarr=CG_rate,\n",
    "\t\t\tsliding_win_len=sliding_win_len,\n",
    "\t\t)\n",
    "\n",
    "\t\tdata = [\n",
    "\t\t\tnet_CG,\n",
    "\t\t\tavg_CG,\n",
    "\t\t\t(pos_counts_C + pos_counts_G),\n",
    "\t\t\tCG_rate,\n",
    "\t\t\tavg_CG_rate,\n",
    "\t\t]\n",
    "\n",
    "\telse:\n",
    "\t\tpositional_GC_count = pos_counts.C.loc[-win_len//2 : win_len//2-1].to_numpy() + pos_counts.G.loc[-win_len//2 : win_len//2-1].to_numpy()\n",
    "\t\tdata = [\n",
    "\t\t\tnet_CG,\n",
    "\t\t\tpositional_GC_count,\n",
    "\t\t]\n",
    "\n",
    "\tdf_tmp = pd.DataFrame(data).T.convert_dtypes()\n",
    "\tdf_tmp.columns = column_headers\n",
    "\tdf_tmp.index = range(-win_len//2, win_len//2)\n",
    "\n",
    "\tif not os.path.exists(parent_dir):\n",
    "\t\tos.makedirs(parent_dir)\n",
    "\n",
    "\tfname = \"netGC\" if not truncated_output else f\"netGC_{win_len}b\"\n",
    "\twith open(f\"{parent_dir}/{fname}.csv\", writing_mode) as fout:\n",
    "\t\tdf_tmp.to_csv(\n",
    "\t\t\tfout,\n",
    "\t\t\tfloat_format='%.10f',\n",
    "\t\t\tlineterminator='\\n',\n",
    "\t\t)\n",
    "\n",
    "\n",
    "def select_pos_counts(ref: str, alt: str, include_CpG: bool, total_pos_counts: pd.DataFrame, nonCpG_pos_counts: pd.DataFrame) -> np.ndarray:\n",
    "    '''\n",
    "    Selects a counts for `ref`, based on `ref` and `alt`, or `include_CpG`.\n",
    "\n",
    "    If `include_CpG` is `True`, will always return `ref` from `total_pos_counts`. Use this for when CpG DNMs\n",
    "    are to be included in positional metrics.\n",
    "    \n",
    "    If `include_CpG` is `False`, `ref` will be returned from `nonCpG_pos_counts` when `ref` is 'C' and `alt` \n",
    "    is 'T', or `ref` is 'G' and `alt` is 'A', otherwise `ref` will be returned from `total_pos_counts`.\n",
    "\n",
    "    ## Parameters\n",
    "    ref: str\n",
    "        Reference genome base.\n",
    "    alt: str\n",
    "        Base found in proband.\n",
    "    include_CpG: bool\n",
    "        Whether to include CpG DNMs in counts.\n",
    "    total_pos_counts: Pandas.DataFrame\n",
    "        DataFrame containing total positional counts (CpG + nonCpG) for `ref`.\n",
    "    nonCpG_pos_counts: Pandas.DataFrame\n",
    "        DataFrame containing nonCpG positional counts for `ref`.\n",
    "    ## Returns\n",
    "    selected_counts: Numpy.NDArray\n",
    "        The selected counts.\n",
    "    '''\n",
    "\n",
    "    if include_CpG or not ((ref == 'C' and alt == 'T') or (ref == 'G' and alt == 'A')):\n",
    "        return total_pos_counts[f'{ref}'].to_numpy(dtype=np.float64)\n",
    "    else:\n",
    "        return nonCpG_pos_counts[f'{ref}'].to_numpy(dtype=np.float64)\n",
    "\n",
    "\n",
    "def get_dnm_counts(df: pd.DataFrame, ref: str, alt: str, win_len: int) -> np.ndarray:\n",
    "    '''\n",
    "    Return positional DNM counts for the DNM `ref` to `alt`.\n",
    "\n",
    "    ## Parameters\n",
    "    df: Pandas.DataFrame\n",
    "        DataFrame containing DNMs dataset.\n",
    "    ref: str\n",
    "        Reference base.\n",
    "    alt: str\n",
    "        Base found in proband.\n",
    "    win_len: int\n",
    "        Sequence window length.\n",
    "    ## Returns\n",
    "    dnm_pos_counts: Numpy.NDArray\n",
    "        Positional DNM counts.\n",
    "    '''\n",
    "\n",
    "    dnm_pos_counts = np.zeros(win_len)\n",
    "\n",
    "    for dnm in df[['REF','ALT','DIST']][(df.REF == ref) & (df.ALT == alt)].values:\n",
    "        dist = dnm[-1]\n",
    "        if dist >= WINDOW_LENGTH//2:\n",
    "            continue\n",
    "        \n",
    "        dnm_pos_counts[dist+win_len//2] += 1\n",
    "\n",
    "    return dnm_pos_counts\n",
    "\n",
    "\n",
    "def sliding_window_avg(arr: np.ndarray, sliding_win_len: int, mode: str='same') -> np.ndarray:\n",
    "    '''\n",
    "    Return sliding window averages over `arr`.\n",
    "\n",
    "    ## Parameters\n",
    "    arr: Numpy.NDArray\n",
    "        DataFrame containing DNMs dataset.\n",
    "    sliding_win_len: int\n",
    "        Sliding window length.\n",
    "    mode: str, default = 'same'\n",
    "        Mode used for Numpy's convolve() function. Default mode is 'same', \n",
    "        which returns an array with the same size as `arr`.\n",
    "    ## Returns\n",
    "    sliding_win_avgs: Numpy.NDArray\n",
    "        Sliding window averages.\n",
    "    '''\n",
    "\n",
    "    # kernel of size sliding_win_len with each element set to 1 / sliding_win_len,\n",
    "    # which behaves like an averaging filter when used as the kernel during convolution\n",
    "    kernel_size = np.ones(sliding_win_len) / sliding_win_len\n",
    "\n",
    "    return np.convolve(\n",
    "        arr,\n",
    "        kernel_size,\n",
    "        mode=mode,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "GC_gain_loss_dnm_types = [\n",
    "\t('A','G'),\n",
    "\t('T','G'),\n",
    "\t('A','C'),\n",
    "\t('T','C'),\n",
    "\t('G','A'),\n",
    "\t('G','T'),\n",
    "\t('C','A'),\n",
    "\t('C','T'),\n",
    "]\n",
    "\n",
    "metrics_columns = [\n",
    "\t\"Number of DNMs\",\n",
    "\t\"Average of DNMs\",\n",
    "\t\"Positional count\",\n",
    "\t\"Rate of mutation\",\n",
    "\t\"Average rate of mutation\",\n",
    "]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **7.1** TSS"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **7.1.1** Total (including CpG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "metadata": {},
   "outputs": [],
   "source": [
    "print_positional_metrics(\n",
    "    df=tss_matches,\n",
    "    total_pos_counts=tss_pos_counts,\n",
    "    nonCpG_pos_counts=tss_nonCpG_pos_counts,\n",
    "    win_len=WINDOW_LENGTH,\n",
    "    sliding_win_len=SLIDING_WIN_LEN,\n",
    "    GC_loss_dnm_types=GC_gain_loss_dnm_types,\n",
    "    include_CpG=True,\n",
    "    column_headers=metrics_columns,\n",
    "    parent_dir=f\"{POSITIONAL_METRICS_SUBDIR}/tss/including_CpG\",\n",
    "    writing_mode=WRITING_MODE,\n",
    ")\n",
    "\n",
    "# net GC\n",
    "print_positional_netGC(\n",
    "    df=tss_matches,\n",
    "    pos_counts=tss_pos_counts,\n",
    "    win_len=WINDOW_LENGTH,\n",
    "    sliding_win_len=SLIDING_WIN_LEN,\n",
    "    GC_loss_dnm_types=GC_gain_loss_dnm_types,\n",
    "    column_headers=metrics_columns,\n",
    "    parent_dir=f\"{POSITIONAL_METRICS_SUBDIR}/tss/including_CpG\",\n",
    "    writing_mode=WRITING_MODE,\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **7.1.2** non-CpG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [],
   "source": [
    "print_positional_metrics(\n",
    "    df=tss_nonCpG,\n",
    "    total_pos_counts=tss_pos_counts,\n",
    "    nonCpG_pos_counts=tss_nonCpG_pos_counts,\n",
    "    win_len=WINDOW_LENGTH,\n",
    "    sliding_win_len=SLIDING_WIN_LEN,\n",
    "    GC_loss_dnm_types=GC_gain_loss_dnm_types,\n",
    "    include_CpG=False,\n",
    "    column_headers=metrics_columns,\n",
    "    parent_dir=f\"{POSITIONAL_METRICS_SUBDIR}/tss/nonCpG\",\n",
    "    writing_mode=WRITING_MODE,\n",
    ")\n",
    "\n",
    "# net GC\n",
    "print_positional_netGC(\n",
    "    df=tss_nonCpG,\n",
    "    pos_counts=tss_nonCpG_pos_counts,\n",
    "    win_len=WINDOW_LENGTH,\n",
    "    sliding_win_len=SLIDING_WIN_LEN,\n",
    "    GC_loss_dnm_types=GC_gain_loss_dnm_types,\n",
    "    column_headers=metrics_columns,\n",
    "    parent_dir=f\"{POSITIONAL_METRICS_SUBDIR}/tss/nonCpG\",\n",
    "    writing_mode=WRITING_MODE,\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **7.2** Intergenic"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **7.2.1** GC matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [],
   "source": [
    "GCmatched_metrics_columns = [\n",
    "    'Number of DNMs',\n",
    "    'Positional count',\n",
    "]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Total (including CpG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "for ref, alt in GC_gain_loss_dnm_types:\n",
    "\traw_tmp = np.zeros(INTERGEN_GCMATCH_WINLEN)\n",
    "\n",
    "\t# dnm count\n",
    "\tfor dnm in intergen_GCmatch_matches[['REF','ALT','DIST']][(intergen_GCmatch_matches.REF == ref) & (intergen_GCmatch_matches.ALT == alt)].values:\n",
    "\t\tdist = dnm[-1]\n",
    "\n",
    "\t\traw_tmp[dist+INTERGEN_GCMATCH_WINLEN//2] += 1\n",
    "\n",
    "\t# aggregation\n",
    "\tdf_tmp = pd.DataFrame(\n",
    "\t\tdata = [\n",
    "\t\t\traw_tmp,\n",
    "\t\t\tintergen_GCmatch_pos_counts[f'{ref}'].loc[-INTERGEN_GCMATCH_WINLEN//2:INTERGEN_GCMATCH_WINLEN//2-1],\n",
    "\t\t],\n",
    "\t).T.convert_dtypes()\n",
    "\tdf_tmp.columns = GCmatched_metrics_columns\n",
    "\tdf_tmp.index = range(-INTERGEN_GCMATCH_WINLEN//2, INTERGEN_GCMATCH_WINLEN//2)\n",
    "\n",
    "\t# write out\n",
    "\twith open(f\"{POSITIONAL_METRICS_SUBDIR}/intergenic/GC_matched/including_CpG/{ref}{alt}_100b.csv\", WRITING_MODE) as fout:\n",
    "\t\tdf_tmp.to_csv(\n",
    "\t\t\tfout,\n",
    "\t\t\tfloat_format='%.10f',\n",
    "\t\t\tlineterminator='\\n',\n",
    "\t\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {},
   "outputs": [],
   "source": [
    "# net GC\n",
    "print_positional_netGC(\n",
    "    df=intergen_GCmatch_matches,\n",
    "    pos_counts=intergen_GCmatch_pos_counts,\n",
    "    win_len=INTERGEN_GCMATCH_WINLEN,\n",
    "    GC_loss_dnm_types=GC_gain_loss_dnm_types,\n",
    "    column_headers=GCmatched_metrics_columns,\n",
    "    intergenic = True,\n",
    "    parent_dir=f\"{POSITIONAL_METRICS_SUBDIR}/intergenic/GC_matched/including_CpG\",\n",
    "    writing_mode=WRITING_MODE,\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### non-CpG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "for ref, alt in GC_gain_loss_dnm_types:\n",
    "\traw_tmp = np.zeros(INTERGEN_GCMATCH_WINLEN)\n",
    "\n",
    "\t# dnm count\n",
    "\tfor dnm in intergen_GCmatch_nonCpG[['REF','ALT','DIST']][(intergen_GCmatch_nonCpG.REF == ref) & (intergen_GCmatch_nonCpG.ALT == alt)].values:\n",
    "\t\t_, _, dist = dnm\n",
    "\n",
    "\t\traw_tmp[dist+INTERGEN_GCMATCH_WINLEN//2] += 1\n",
    "\n",
    "\t# aggregation\n",
    "\tdf_tmp = pd.DataFrame(\n",
    "\t\tdata = [\n",
    "\t\t\traw_tmp,\n",
    "\t\t\tintergen_GCmatch_nonCpG_pos_counts[f'{ref}'].loc[-INTERGEN_GCMATCH_WINLEN//2:INTERGEN_GCMATCH_WINLEN//2-1].to_numpy(dtype=np.float64) if (ref == 'C' and alt == 'T') or (ref == 'G' and alt == 'A') else intergen_GCmatch_pos_counts[f'{ref}'].loc[-INTERGEN_GCMATCH_WINLEN//2:INTERGEN_GCMATCH_WINLEN//2-1].to_numpy(dtype=np.float64),\n",
    "\t\t],\n",
    "\t).T.convert_dtypes()\n",
    "\tdf_tmp.columns = ['Number of DNMs', 'Positional count']\n",
    "\tdf_tmp.index = range(-INTERGEN_GCMATCH_WINLEN//2, INTERGEN_GCMATCH_WINLEN//2)\n",
    "\n",
    "\t# write out\n",
    "\twith open(f\"{POSITIONAL_METRICS_SUBDIR}/intergenic/GC_matched/non_CpG/{ref}{alt}_100b.csv\", WRITING_MODE) as fout:\n",
    "\t\tdf_tmp.to_csv(\n",
    "\t\t\tfout,\n",
    "\t\t\tfloat_format='%.10f',\n",
    "\t\t\tlineterminator='\\n',\n",
    "\t\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {},
   "outputs": [],
   "source": [
    "# net GC\n",
    "\n",
    "print_positional_netGC(\n",
    "    df=intergen_GCmatch_nonCpG,\n",
    "    pos_counts=intergen_GCmatch_nonCpG_pos_counts,\n",
    "    win_len=INTERGEN_GCMATCH_WINLEN,\n",
    "    GC_loss_dnm_types=GC_gain_loss_dnm_types,\n",
    "    column_headers=GCmatched_metrics_columns,\n",
    "    intergenic = True,\n",
    "    parent_dir=f\"{POSITIONAL_METRICS_SUBDIR}/intergenic/GC_matched/non_CpG\",\n",
    "    writing_mode=WRITING_MODE,\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **7.2.2** non-GC matched"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Total (including CpG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 302,
   "metadata": {},
   "outputs": [],
   "source": [
    "print_positional_metrics(\n",
    "    df=intergen_nonGCmatch_matches,\n",
    "    total_pos_counts=intergen_nonGCmatch_pos_counts,\n",
    "    nonCpG_pos_counts=intergen_nonGCmatch_nonCpG_pos_counts,\n",
    "    win_len=WINDOW_LENGTH,\n",
    "    sliding_win_len=SLIDING_WIN_LEN,\n",
    "    GC_loss_dnm_types=GC_gain_loss_dnm_types,\n",
    "    include_CpG=True,\n",
    "    column_headers=metrics_columns,\n",
    "    parent_dir=f\"{POSITIONAL_METRICS_SUBDIR}/intergenic/non_GC_matched/including_CpG\",\n",
    "    writing_mode=WRITING_MODE,\n",
    ")\n",
    "\n",
    "# net GC\n",
    "print_positional_netGC(\n",
    "    df=intergen_nonGCmatch_matches,\n",
    "    pos_counts=intergen_nonGCmatch_pos_counts,\n",
    "    win_len=WINDOW_LENGTH,\n",
    "    sliding_win_len=SLIDING_WIN_LEN,\n",
    "    GC_loss_dnm_types=GC_gain_loss_dnm_types,\n",
    "    column_headers=metrics_columns,\n",
    "    parent_dir=f\"{POSITIONAL_METRICS_SUBDIR}/intergenic/non_GC_matched/including_CpG\",\n",
    "    writing_mode=WRITING_MODE,\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### non-CpG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 303,
   "metadata": {},
   "outputs": [],
   "source": [
    "print_positional_metrics(\n",
    "    df=intergen_nonGCmatch_nonCpG,\n",
    "    total_pos_counts=intergen_nonGCmatch_pos_counts,\n",
    "    nonCpG_pos_counts=intergen_nonGCmatch_nonCpG_pos_counts,\n",
    "    win_len=WINDOW_LENGTH,\n",
    "    sliding_win_len=SLIDING_WIN_LEN,\n",
    "    GC_loss_dnm_types=GC_gain_loss_dnm_types,\n",
    "    include_CpG=False,\n",
    "    column_headers=metrics_columns,\n",
    "    parent_dir=f\"{POSITIONAL_METRICS_SUBDIR}/intergenic/non_GC_matched/non_CpG\",\n",
    "    writing_mode=WRITING_MODE,\n",
    ")\n",
    "\n",
    "# net GC\n",
    "print_positional_netGC(\n",
    "    df=intergen_nonGCmatch_nonCpG,\n",
    "    pos_counts=intergen_nonGCmatch_nonCpG_pos_counts,\n",
    "    win_len=WINDOW_LENGTH,\n",
    "    sliding_win_len=SLIDING_WIN_LEN,\n",
    "    GC_loss_dnm_types=GC_gain_loss_dnm_types,\n",
    "    column_headers=metrics_columns,\n",
    "    parent_dir=f\"{POSITIONAL_METRICS_SUBDIR}/intergenic/non_GC_matched/non_CpG\",\n",
    "    writing_mode=WRITING_MODE,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "________________"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **8.** <u>Gene Level Metrics</u>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_identifiers = tss_windows.ENSGID.to_list()\n",
    "\n",
    "# # write out Ensembl gene IDs (optional)\n",
    "# gene_identifiers_path = f\"{COORDS_DIR}/hg38_genes.csv\"\n",
    "# with open(gene_identifiers_path, WRITING_MODE) as fout:\n",
    "#     tss_windows.ENSGID.to_csv(\n",
    "#         fout,\n",
    "#         lineterminator='\\n',\n",
    "#         index=False,\n",
    "#         header=False,\n",
    "#     )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **8.1** TSS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **8.1.1** All TSS DNMs (CpG + non-CpG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_gene_DNM_distances_dict = { gene_id : dict(zip(DNM_TYPES,[list() for i in range(len(DNM_TYPES))]))\n",
    "\t\t\t\t\t\t\t\tfor gene_id\n",
    "\t\t\t\t\t\t\t\tin gene_identifiers }\n",
    "\n",
    "for gene_id, dnm_distances in tss_gene_DNM_distances_dict.items():\n",
    "\tfor dnm in tss_matches[tss_matches.ENSGID == gene_id].values:\n",
    "\t\tchrom, pos, ref, alt, tss, eib, dist, first_ex, downstream, intergenic, strand, ensgid, enstid, window = dnm\n",
    "\n",
    "\t\tdnm_distances['>'.join([ref, alt])].append(dist+WINDOW_LENGTH//2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A&gt;G</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>A&gt;T</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>A&gt;C</th>\n",
       "      <th>G&gt;A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000187634</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188976</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187961</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187583</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188290</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000129862</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000165246</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000012817</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198692</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000280969</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17104 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                A>G T>C C>G T>G C>A A>T G>C G>T C>T T>A A>C G>A\n",
       "ENSG00000187634  []  []  []  []  []  []  []  []  []  []  []  []\n",
       "ENSG00000188976  []  []  []  []  []  []  []  []  []  []  []  []\n",
       "ENSG00000187961  []  []  []  []  []  []  []  []  []  []  []  []\n",
       "ENSG00000187583  []  []  []  []  []  []  []  []  []  []  []  []\n",
       "ENSG00000188290  []  []  []  []  []  []  []  []  []  []  []  []\n",
       "...              ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..\n",
       "ENSG00000129862  []  []  []  []  []  []  []  []  []  []  []  []\n",
       "ENSG00000165246  []  []  []  []  []  []  []  []  []  []  []  []\n",
       "ENSG00000012817  []  []  []  []  []  []  []  []  []  []  []  []\n",
       "ENSG00000198692  []  []  []  []  []  []  []  []  []  []  []  []\n",
       "ENSG00000280969  []  []  []  []  []  []  []  []  []  []  []  []\n",
       "\n",
       "[17104 rows x 12 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tss_gene_DNM_distances = pd.DataFrame(\n",
    "\ttss_gene_DNM_distances_dict,\n",
    ").T\n",
    "\n",
    "tss_gene_DNM_distances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_gene_DNM_distances_repr = tss_gene_DNM_distances.astype(str).apply(lambda x: x.apply(lambda y: y.strip('[]')))\n",
    "\n",
    "tss_gene_DNM_distances_repr = tss_gene_DNM_distances_repr[[\n",
    "\t'A>T',\n",
    "\t'A>G',\n",
    "\t'A>C',\n",
    "\t'T>A',\n",
    "\t'T>G',\n",
    "\t'T>C',\n",
    "\t'G>A',\n",
    "\t'G>T',\n",
    "\t'G>C',\n",
    "\t'C>A',\n",
    "\t'C>T',\n",
    "\t'C>G',\n",
    "]]\n",
    "\n",
    "# write out results\n",
    "with open(f\"{GENE_LEVEL_METRICS_SUBDIR}/tss/tss_all.csv\", WRITING_MODE) as fout:\n",
    "\ttss_gene_DNM_distances_repr.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **8.1.2** Split CpG and non-CpG TSS DNMs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_gene_DNM_distances_CpG_dict = { gene_id : dict(zip(DNM_TYPES,[list() for i in range(len(DNM_TYPES))]))\n",
    "\t\t\t\t\t\t\t\t\tfor gene_id\n",
    "\t\t\t\t\t\t\t\t\tin gene_identifiers }\n",
    "tss_gene_DNM_distances_nonCpG_dict = { gene_id : dict(zip(DNM_TYPES,[list() for i in range(len(DNM_TYPES))]))\n",
    "\t\t\t\t\t\t\t\t\t\tfor gene_id\n",
    "\t\t\t\t\t\t\t\t\t\tin gene_identifiers }\n",
    "\n",
    "for gene_id in gene_identifiers:\n",
    "\tfor dnm in tss_matches[tss_matches.ENSGID == gene_id].values:\n",
    "\t\tchrom, pos, ref, alt, tss, eib, dist, first_ex, downstream, intergenic, strand, ensgid, enstid, window = dnm\n",
    "\n",
    "\t\tis_CpG = (\n",
    "\t\t\t(ref == 'C' and window[WINDOW_LENGTH//2+int(dist)+1+CPG_OFFSET] == 'G') or\n",
    "\t\t\t(ref == 'G' and window[WINDOW_LENGTH//2+int(dist)-1+CPG_OFFSET] == 'C')\n",
    "\t\t)\n",
    "\n",
    "\t\tif is_CpG:\n",
    "\t\t\ttss_gene_DNM_distances_CpG_dict[gene_id]['>'.join([ref, alt])].append(dist+WINDOW_LENGTH//2)\n",
    "\t\telse:\n",
    "\t\t\ttss_gene_DNM_distances_nonCpG_dict[gene_id]['>'.join([ref, alt])].append(dist+WINDOW_LENGTH//2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_gene_DNM_distances_CpG = pd.DataFrame(\n",
    "\ttss_gene_DNM_distances_CpG_dict,\n",
    ").T\n",
    "\n",
    "tss_gene_DNM_distances_nonCpG = pd.DataFrame(\n",
    "\ttss_gene_DNM_distances_nonCpG_dict,\n",
    ").T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_gene_DNM_distances_CpG_repr = tss_gene_DNM_distances_CpG.astype(str).apply(lambda x: x.apply(lambda y: y.strip('[]')))\n",
    "\n",
    "tss_gene_DNM_distances_CpG_repr = tss_gene_DNM_distances_CpG_repr[[\n",
    "\t'G>A',\n",
    "\t'G>T',\n",
    "\t'G>C',\n",
    "\t'C>A',\n",
    "\t'C>T',\n",
    "\t'C>G',\n",
    "]]\n",
    "\n",
    "with open(f\"{GENE_LEVEL_METRICS_SUBDIR}/tss/tss_CpG.csv\", WRITING_MODE) as fout:\n",
    "\ttss_gene_DNM_distances_CpG_repr.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_gene_DNM_distances_nonCpG_repr = tss_gene_DNM_distances_nonCpG.astype(str).apply(lambda x: x.apply(lambda y: y.strip('[]')))\n",
    "\n",
    "tss_gene_DNM_distances_nonCpG_repr = tss_gene_DNM_distances_nonCpG_repr[[\n",
    "\t'A>T',\n",
    "\t'A>G',\n",
    "\t'A>C',\n",
    "\t'T>A',\n",
    "\t'T>G',\n",
    "\t'T>C',\n",
    "\t'G>A',\n",
    "\t'G>T',\n",
    "\t'G>C',\n",
    "\t'C>A',\n",
    "\t'C>T',\n",
    "\t'C>G',\n",
    "]]\n",
    "\n",
    "with open(f\"{GENE_LEVEL_METRICS_SUBDIR}/tss/tss_nonCpG.csv\", WRITING_MODE) as fout:\n",
    "\ttss_gene_DNM_distances_nonCpG_repr.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **8.2** Intergenic GC matched"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **8.2.1** All intergenic GC matched DNMs (CpG + non-CpG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergen_GCmatch_DNM_distances_dict = { f\"chr{chrom}:{center-INTERGEN_GCMATCH_WINLEN//2}-{center+INTERGEN_GCMATCH_WINLEN//2}\" : dict(zip(DNM_TYPES,[list() for i in range(len(DNM_TYPES))]))\n",
    "\t\t\t\t\t\t\t\t\t\t\tfor chrom, center, _\n",
    "\t\t\t\t\t\t\t\t\t\t\tin intergen_GCmatch_windows.values }\n",
    "\n",
    "for gene_id, dnm_distances in intergen_GCmatch_DNM_distances_dict.items():\n",
    "\tref_chrom, start, end = re.split('[:-]+',gene_id)\n",
    "\tref_chrom = re.findall(r'\\d+',ref_chrom)[0]\n",
    "\n",
    "\ttemp_matches = intergen_GCmatch_matches[\n",
    "\t\t(intergen_GCmatch_matches.CHROM.astype(str) == ref_chrom) &\n",
    "\t\t(intergen_GCmatch_matches.CENTER == (int(start) + INTERGEN_GCMATCH_WINLEN//2))\n",
    "\t]\n",
    "\n",
    "\tfor dnm in temp_matches.values:\n",
    "\t\tdnm_chrom, pos, ref, alt, center, dist, window = dnm\n",
    "\n",
    "\t\tdnm_distances['>'.join([ref, alt])].append(dist+WINDOW_LENGTH//2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A&gt;G</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>A&gt;T</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>A&gt;C</th>\n",
       "      <th>G&gt;A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:114327737-114327837</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:114328137-114328237</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:114328237-114328337</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:47396304-47396404</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:166420304-166420404</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr21:8439401-8439501</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22:44331786-44331886</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22:46071211-46071311</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22:16601785-16601885</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22:48086859-48086959</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16568 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         A>G T>C C>G T>G C>A A>T G>C G>T C>T T>A A>C G>A\n",
       "chr1:114327737-114327837  []  []  []  []  []  []  []  []  []  []  []  []\n",
       "chr1:114328137-114328237  []  []  []  []  []  []  []  []  []  []  []  []\n",
       "chr1:114328237-114328337  []  []  []  []  []  []  []  []  []  []  []  []\n",
       "chr1:47396304-47396404    []  []  []  []  []  []  []  []  []  []  []  []\n",
       "chr1:166420304-166420404  []  []  []  []  []  []  []  []  []  []  []  []\n",
       "...                       ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..\n",
       "chr21:8439401-8439501     []  []  []  []  []  []  []  []  []  []  []  []\n",
       "chr22:44331786-44331886   []  []  []  []  []  []  []  []  []  []  []  []\n",
       "chr22:46071211-46071311   []  []  []  []  []  []  []  []  []  []  []  []\n",
       "chr22:16601785-16601885   []  []  []  []  []  []  []  []  []  []  []  []\n",
       "chr22:48086859-48086959   []  []  []  []  []  []  []  []  []  []  []  []\n",
       "\n",
       "[16568 rows x 12 columns]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergen_GCmatch_DNM_distances = pd.DataFrame(\n",
    "\tintergen_GCmatch_DNM_distances_dict,\n",
    ").T\n",
    "\n",
    "intergen_GCmatch_DNM_distances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergen_GCmatch_DNM_distances_repr = intergen_GCmatch_DNM_distances.astype(str).apply(lambda x: x.apply(lambda y: y.strip('[]')))\n",
    "\n",
    "intergen_GCmatch_DNM_distances_repr = intergen_GCmatch_DNM_distances_repr[[\n",
    "\t'A>T',\n",
    "\t'A>G',\n",
    "\t'A>C',\n",
    "\t'T>A',\n",
    "\t'T>G',\n",
    "\t'T>C',\n",
    "\t'G>A',\n",
    "\t'G>T',\n",
    "\t'G>C',\n",
    "\t'C>A',\n",
    "\t'C>T',\n",
    "\t'C>G',\n",
    "]]\n",
    "\n",
    "with open(f\"{GENE_LEVEL_METRICS_SUBDIR}/intergen/GCmatch/intergen_GCmatch_all_100b.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_GCmatch_DNM_distances_repr.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **8.2.2** Split CpG and non-CpG intergenic GC matched DNMs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergen_GCmatch_DNM_distances_CpG_dict = { gene_id : dict(zip(DNM_TYPES,[list() for i in range(len(DNM_TYPES))]))\n",
    "\t\t\t\t\t\t\t\t\t\t\tfor gene_id\n",
    "\t\t\t\t\t\t\t\t\t\t\tin intergen_GCmatch_DNM_distances_dict.keys() }\n",
    "intergen_GCmatch_DNM_distances_nonCpG_dict = { gene_id : dict(zip(DNM_TYPES,[list() for i in range(len(DNM_TYPES))]))\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tfor gene_id\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tin intergen_GCmatch_DNM_distances_dict.keys() }\n",
    "\n",
    "\n",
    "for gene_id in intergen_GCmatch_DNM_distances_dict.keys():\n",
    "\tref_chrom, start, _ = re.split('[:-]+',gene_id)\n",
    "\tref_chrom = re.findall(r'\\d+',ref_chrom)[0]\n",
    "\n",
    "\ttemp_matches = intergen_GCmatch_matches[\n",
    "\t\t(intergen_GCmatch_matches.CHROM.astype(str) == ref_chrom) &\n",
    "\t\t(intergen_GCmatch_matches.CENTER == (int(start) + INTERGEN_GCMATCH_WINLEN//2))\n",
    "\t]\n",
    "\n",
    "\tfor dnm in temp_matches.values:\n",
    "\t\t_, _, ref, alt, _, dist, window = dnm\n",
    "\n",
    "\t\tis_CpG = (\n",
    "\t\t\t(ref == 'C' and window[WINDOW_LENGTH//2+int(dist)+1+CPG_OFFSET*0] == 'G') or\n",
    "\t\t\t(ref == 'G' and window[WINDOW_LENGTH//2+int(dist)-1+CPG_OFFSET*0] == 'C')\n",
    "\t\t)\n",
    "\n",
    "\t\tif is_CpG:\n",
    "\t\t\tintergen_GCmatch_DNM_distances_CpG_dict[gene_id]['>'.join([ref, alt])].append(dist+WINDOW_LENGTH//2)\n",
    "\t\telse:\n",
    "\t\t\tintergen_GCmatch_DNM_distances_nonCpG_dict[gene_id]['>'.join([ref, alt])].append(dist+WINDOW_LENGTH//2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergen_GCmatch_DNM_distances_CpG = pd.DataFrame(\n",
    "\tintergen_GCmatch_DNM_distances_CpG_dict,\n",
    ").T\n",
    "\n",
    "intergen_GCmatch_DNM_distances_nonCpG = pd.DataFrame(\n",
    "\tintergen_GCmatch_DNM_distances_nonCpG_dict,\n",
    ").T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergen_GCmatch_DNM_distances_CpG_repr = intergen_GCmatch_DNM_distances_CpG.astype(str).apply(lambda x: x.apply(lambda y: y.strip('[]')))\n",
    "\n",
    "intergen_GCmatch_DNM_distances_CpG_repr = intergen_GCmatch_DNM_distances_CpG_repr[[\n",
    "\t'G>A',\n",
    "\t'G>T',\n",
    "\t'G>C',\n",
    "\t'C>A',\n",
    "\t'C>T',\n",
    "\t'C>G',\n",
    "]]\n",
    "\n",
    "with open(f\"{GENE_LEVEL_METRICS_SUBDIR}/intergen/GCmatch/intergen_GCmatch_CpG_100b.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_GCmatch_DNM_distances_CpG_repr.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergen_GCmatch_DNM_distances_nonCpG_repr = intergen_GCmatch_DNM_distances_nonCpG.astype(str).apply(lambda x: x.apply(lambda y: y.strip('[]')))\n",
    "\n",
    "intergen_GCmatch_DNM_distances_nonCpG_repr = intergen_GCmatch_DNM_distances_nonCpG_repr[[\n",
    "\t'A>T',\n",
    "\t'A>G',\n",
    "\t'A>C',\n",
    "\t'T>A',\n",
    "\t'T>G',\n",
    "\t'T>C',\n",
    "\t'G>A',\n",
    "\t'G>T',\n",
    "\t'G>C',\n",
    "\t'C>A',\n",
    "\t'C>T',\n",
    "\t'C>G',\n",
    "]]\n",
    "\n",
    "with open(f\"{GENE_LEVEL_METRICS_SUBDIR}/intergen/GCmatch/intergen_GCmatch_nonCpG_100b.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_GCmatch_DNM_distances_nonCpG_repr.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **8.3** Intergenic non-GC matched"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **8.3.1** All intergenic non-GC matched DNMs (CpG + non-CpG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergen_nonGCmatch_DNM_distances_dict = { f\"chr{chrom}:{center-WINDOW_LENGTH//2}-{center+WINDOW_LENGTH//2}\" : dict(zip(DNM_TYPES,[list() for i in range(len(DNM_TYPES))]))\n",
    "\t\t\t\t\t\t\t\t\t\t\tfor chrom, center, window\n",
    "\t\t\t\t\t\t\t\t\t\t\tin intergen_nonGCmatch_windows.values }\n",
    "\n",
    "for gene_id, dnm_distances in intergen_nonGCmatch_DNM_distances_dict.items():\n",
    "\tref_chrom, start, end = re.split('[:-]+',gene_id)\n",
    "\tref_chrom = re.findall(r'\\d+',ref_chrom)[0]\n",
    "\n",
    "\ttemp_matches = intergen_nonGCmatch_matches[\n",
    "\t\t(intergen_nonGCmatch_matches.CHROM.astype(str) == ref_chrom) &\n",
    "\t\t(intergen_nonGCmatch_matches.CENTER == (int(start) + WINDOW_LENGTH//2))\n",
    "\t]\n",
    "\n",
    "\tfor dnm in temp_matches.values:\n",
    "\t\tdnm_chrom, pos, ref, alt, center, dist, window = dnm\n",
    "\n",
    "\t\tdnm_distances['>'.join([ref, alt])].append(dist+WINDOW_LENGTH//2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A&gt;G</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>A&gt;T</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>A&gt;C</th>\n",
       "      <th>G&gt;A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1:114326537-114331537</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:47396104-47401104</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:166419304-166424304</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:69212604-69217604</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[3671]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1:29409919-29414919</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22:11484544-11489544</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22:27191025-27196025</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22:12061784-12066784</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22:29301107-29306107</th>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[3895]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22:30630117-30635117</th>\n",
       "      <td>[2561]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163508 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             A>G T>C C>G T>G C>A     A>T G>C G>T C>T T>A A>C  \\\n",
       "chr1:114326537-114331537      []  []  []  []  []      []  []  []  []  []  []   \n",
       "chr1:47396104-47401104        []  []  []  []  []      []  []  []  []  []  []   \n",
       "chr1:166419304-166424304      []  []  []  []  []      []  []  []  []  []  []   \n",
       "chr1:69212604-69217604        []  []  []  []  []  [3671]  []  []  []  []  []   \n",
       "chr1:29409919-29414919        []  []  []  []  []      []  []  []  []  []  []   \n",
       "...                          ...  ..  ..  ..  ..     ...  ..  ..  ..  ..  ..   \n",
       "chr22:11484544-11489544       []  []  []  []  []      []  []  []  []  []  []   \n",
       "chr22:27191025-27196025       []  []  []  []  []      []  []  []  []  []  []   \n",
       "chr22:12061784-12066784       []  []  []  []  []      []  []  []  []  []  []   \n",
       "chr22:29301107-29306107       []  []  []  []  []      []  []  []  []  []  []   \n",
       "chr22:30630117-30635117   [2561]  []  []  []  []      []  []  []  []  []  []   \n",
       "\n",
       "                             G>A  \n",
       "chr1:114326537-114331537      []  \n",
       "chr1:47396104-47401104        []  \n",
       "chr1:166419304-166424304      []  \n",
       "chr1:69212604-69217604        []  \n",
       "chr1:29409919-29414919        []  \n",
       "...                          ...  \n",
       "chr22:11484544-11489544       []  \n",
       "chr22:27191025-27196025       []  \n",
       "chr22:12061784-12066784       []  \n",
       "chr22:29301107-29306107   [3895]  \n",
       "chr22:30630117-30635117       []  \n",
       "\n",
       "[163508 rows x 12 columns]"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intergen_nonGCmatch_DNM_distances = pd.DataFrame(\n",
    "\tintergen_nonGCmatch_DNM_distances_dict,\n",
    ").T\n",
    "\n",
    "intergen_nonGCmatch_DNM_distances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergen_nonGCmatch_DNM_distances_repr = intergen_nonGCmatch_DNM_distances.astype(str).apply(lambda x: x.apply(lambda y: y.strip('[]')))\n",
    "\n",
    "intergen_nonGCmatch_DNM_distances_repr = intergen_nonGCmatch_DNM_distances_repr[[\n",
    "\t'A>T',\n",
    "\t'A>G',\n",
    "\t'A>C',\n",
    "\t'T>A',\n",
    "\t'T>G',\n",
    "\t'T>C',\n",
    "\t'G>A',\n",
    "\t'G>T',\n",
    "\t'G>C',\n",
    "\t'C>A',\n",
    "\t'C>T',\n",
    "\t'C>G',\n",
    "]]\n",
    "\n",
    "with open(f\"{GENE_LEVEL_METRICS_SUBDIR}/intergen/nonGCmatch/intergen_nonGCmatch_all.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_nonGCmatch_DNM_distances_repr.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **8.3.2** Split CpG and non-CpG intergenic non-GC matched DNMs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergen_nonGCmatch_DNM_distances_CpG_dict = { gene_id : dict(zip(DNM_TYPES,[list() for i in range(len(DNM_TYPES))]))\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tfor gene_id\n",
    "\t\t\t\t\t\t\t\t\t\t\t\tin intergen_nonGCmatch_DNM_distances_dict.keys() }\n",
    "intergen_nonGCmatch_DNM_distances_nonCpG_dict = { gene_id : dict(zip(DNM_TYPES,[list() for i in range(len(DNM_TYPES))]))\n",
    "\t\t\t\t\t\t\t\t\t\t\t\t\tfor gene_id\n",
    "\t\t\t\t\t\t\t\t\t\t\t\t\tin intergen_nonGCmatch_DNM_distances_dict.keys() }\n",
    "\n",
    "\n",
    "for gene_id in intergen_nonGCmatch_DNM_distances_dict.keys():\n",
    "\tref_chrom, start, _ = re.split('[:-]+',gene_id)\n",
    "\tref_chrom = re.findall(r'\\d+',ref_chrom)[0]\n",
    "\n",
    "\ttemp_matches = intergen_nonGCmatch_matches[\n",
    "\t\t(intergen_nonGCmatch_matches.CHROM.astype(str) == ref_chrom) &\n",
    "\t\t(intergen_nonGCmatch_matches.CENTER == (int(start) + WINDOW_LENGTH//2))\n",
    "\t]\n",
    "\n",
    "\tfor dnm in temp_matches.values:\n",
    "\t\t_, _, ref, alt, _, dist, window = dnm\n",
    "\n",
    "\t\tis_CpG = (\n",
    "\t\t\t(ref == 'C' and window[WINDOW_LENGTH//2+int(dist)+1+CPG_OFFSET*0] == 'G') or\n",
    "\t\t\t(ref == 'G' and window[WINDOW_LENGTH//2+int(dist)-1+CPG_OFFSET*0] == 'C')\n",
    "\t\t)\n",
    "\n",
    "\t\tif is_CpG:\n",
    "\t\t\tintergen_nonGCmatch_DNM_distances_CpG_dict[gene_id]['>'.join([ref, alt])].append(dist+WINDOW_LENGTH//2)\n",
    "\t\telse:\n",
    "\t\t\tintergen_nonGCmatch_DNM_distances_nonCpG_dict[gene_id]['>'.join([ref, alt])].append(dist+WINDOW_LENGTH//2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergen_nonGCmatch_DNM_distances_CpG = pd.DataFrame(\n",
    "\tintergen_nonGCmatch_DNM_distances_CpG_dict,\n",
    ").T\n",
    "\n",
    "intergen_nonGCmatch_DNM_distances_nonCpG = pd.DataFrame(\n",
    "\tintergen_nonGCmatch_DNM_distances_nonCpG_dict,\n",
    ").T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergen_nonGCmatch_DNM_distances_CpG_repr = intergen_nonGCmatch_DNM_distances_CpG.astype(str).apply(lambda x: x.apply(lambda y: y.strip('[]')))\n",
    "\n",
    "intergen_nonGCmatch_DNM_distances_CpG_repr = intergen_nonGCmatch_DNM_distances_CpG_repr[[\n",
    "\t'G>A',\n",
    "\t'G>T',\n",
    "\t'G>C',\n",
    "\t'C>A',\n",
    "\t'C>T',\n",
    "\t'C>G',\n",
    "]]\n",
    "\n",
    "with open(f\"{GENE_LEVEL_METRICS_SUBDIR}/intergen/nonGCmatch/intergen_nonGCmatch_CpG.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_nonGCmatch_DNM_distances_CpG_repr.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergen_nonGCmatch_DNM_distances_nonCpG_repr = intergen_nonGCmatch_DNM_distances_nonCpG.astype(str).apply(lambda x: x.apply(lambda y: y.strip('[]')))\n",
    "\n",
    "intergen_nonGCmatch_DNM_distances_nonCpG_repr = intergen_nonGCmatch_DNM_distances_nonCpG_repr[[\n",
    "\t'A>T',\n",
    "\t'A>G',\n",
    "\t'A>C',\n",
    "\t'T>A',\n",
    "\t'T>G',\n",
    "\t'T>C',\n",
    "\t'G>A',\n",
    "\t'G>T',\n",
    "\t'G>C',\n",
    "\t'C>A',\n",
    "\t'C>T',\n",
    "\t'C>G',\n",
    "]]\n",
    "\n",
    "with open(f\"{GENE_LEVEL_METRICS_SUBDIR}/intergen/nonGCmatch/intergen_nonGCmatch_nonCpG.csv\", WRITING_MODE) as fout:\n",
    "\tintergen_nonGCmatch_DNM_distances_nonCpG_repr.to_csv(\n",
    "\t\tfout,\n",
    "\t\tlineterminator='\\n',\n",
    "\t)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
